// binance-api-node/dist/http-client.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.deliveryCandleFields = exports.default = exports.candleFields = void 0;
var _lodash = _interopRequireDefault(require("lodash.zipobject"));
var _jsonBigint = _interopRequireDefault(require("json-bigint"));
var _signature = require("./signature");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
// Robust environment detection for Node.js vs Browser
var isNode = function () {
  // Check for Node.js specific features
  if (typeof process !== 'undefined' && process.versions !== null && process.versions.node !== null) {
    return true;
  }
  // Check for Deno
  /* eslint-disable no-undef */
  if (typeof Deno !== 'undefined' && Deno.version !== null) {
    return true;
  }
  /* eslint-enable no-undef */
  // Browser or Web Worker
  return false;
}();

// Platform-specific imports
var fetch;
var HttpsProxyAgent;
if (isNode) {
  // Node.js environment - use node-fetch for proper proxy support
  var nodeFetch = require('node-fetch');
  fetch = nodeFetch.default || nodeFetch;
  var proxyAgent = require('https-proxy-agent');
  HttpsProxyAgent = proxyAgent.HttpsProxyAgent || proxyAgent.default || proxyAgent;
} else {
  var _globalThis$fetch, _window$fetch;
  // Browser environment - use native APIs
  /* eslint-disable no-undef */
  fetch = ((_globalThis$fetch = globalThis.fetch) === null || _globalThis$fetch === void 0 ? void 0 : _globalThis$fetch.bind(globalThis)) || ((_window$fetch = window.fetch) === null || _window$fetch === void 0 ? void 0 : _window$fetch.bind(window));
  /* eslint-enable no-undef */
}
var getEndpoint = function getEndpoint(endpoints, path, testnet) {
  var isFutures = path.includes('/fapi') || path.includes('/futures');
  var isDelivery = path.includes('/dapi');
  var isPortfolioMargin = path.includes('/papi');
  if (testnet) {
    if (isFutures) return 'https://demo-fapi.binance.com';
    if (isDelivery) return 'https://demo-dapi.binance.com';
    if (isPortfolioMargin) return 'https://demo-papi.binance.com';
    return 'https://demo-api.binance.com';
  }
  if (path.includes('/fapi') || path.includes('/futures')) return endpoints.futures || 'https://fapi.binance.com';
  if (path.includes('/dapi')) return endpoints.delivery || 'https://dapi.binance.com';
  if (path.includes('/papi')) return endpoints.portfolioMargin || 'https://papi.binance.com';
  return endpoints.base || 'https://api.binance.com';
};
var getDomainName = function getDomainName(url) {
  var match = url.match(/^(?:https?:\/\/)?(?:www\.)?([\w-]+(\.[\w-]+)+)/);
  return match ? match[1] : null;
};
var defaultGetTime = function defaultGetTime() {
  return Date.now();
};
var _uuid = function uuid22(a) {
  return a ? (a ^ Math.random() * 16 >> a / 4).toString(16) : ([1e7] + 1e3 + 4e3 + 8e5).replace(/[018]/g, _uuid);
};

// Singleton holding header data like rate limits.
var info = {};

/**
 * Build query string for uri encoded url based on json object
 */
var makeQueryString = function makeQueryString(q) {
  return q ? "?".concat(Object.keys(q).map(function (k) {
    return "".concat(encodeURIComponent(k), "=").concat(encodeURIComponent(q[k]));
  }).join('&')) : '';
};

/**
 * Get API limits info from headers
 */
var headersMapping = {
  'x-mbx-used-weight-1m': 'usedWeight1m',
  'x-mbx-order-count-10s': 'orderCount10s',
  'x-mbx-order-count-1m': 'orderCount1m',
  'x-mbx-order-count-1h': 'orderCount1h',
  'x-mbx-order-count-1d': 'orderCount1d',
  'x-response-time': 'responseTime'
};
var responseHandler = function responseHandler(res) {
  if (!res.headers || !res.url) return;
  var domain = getDomainName(res.url);
  if (!info[domain]) info[domain] = {};
  for (var _i = 0, _Object$keys = Object.keys(headersMapping); _i < _Object$keys.length; _i++) {
    var key = _Object$keys[_i];
    var outKey = headersMapping[key];
    if (res.headers.has(key)) {
      info[domain][outKey] = res.headers.get(key);
    }
  }
};

/**
 * Finalize API response
 */
var sendResult = function sendResult(call) {
  return call.then(function (res) {
    // Get API limits info from headers
    responseHandler(res);

    // If response is ok, we can safely assume it is valid JSON
    if (res.ok) return res.text().then(function (text) {
      return _jsonBigint.default.parse(text);
    });

    // Errors might come from the API itself or the proxy Binance is using.
    // For API errors the response will be valid JSON,but for proxy errors
    // it will be HTML
    return res.text().then(function (text) {
      var error;
      try {
        var json = _jsonBigint.default.parse(text);
        // The body was JSON parseable, assume it is an API response error
        error = new Error(json.msg || "".concat(res.status, " ").concat(res.statusText));
        error.code = json.code;
        error.url = res.url;
      } catch (e) {
        // The body was not JSON parseable, assume it is proxy error
        error = new Error("".concat(res.status, " ").concat(res.statusText, " ").concat(text));
        error.response = res;
        error.responseText = text;
      }
      throw error;
    });
  });
};

/**
 * Util to validate existence of required parameter(s)
 */
var checkParams = function checkParams(name, payload) {
  var requires = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
  if (!payload) {
    throw new Error('You need to pass a payload object.');
  }
  requires.forEach(function (r) {
    if (!payload[r] && isNaN(payload[r])) {
      throw new Error("Method ".concat(name, " requires ").concat(r, " parameter."));
    }
  });
  return true;
};
var spotP = function spotP() {
  return "x-B3AUXNYV".concat(_uuid());
};
var futuresP = function futuresP() {
  return "x-ftGmvgAN".concat(_uuid());
};

/**
 * Make public calls against the api
 *
 * @param {string} path Endpoint path
 * @param {object} data The payload to be sent
 * @param {string} method HTTB VERB, GET by default
 * @param {object} headers
 * @returns {object} The api response
 */
var publicCall = function publicCall(_ref) {
  var proxy = _ref.proxy,
    endpoints = _ref.endpoints,
    testnet = _ref.testnet;
  return function (path, data) {
    var method = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'GET';
    var headers = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {};
    var fetchOptions = {
      method: method,
      json: true,
      headers: headers
    };

    // Only add proxy agent in Node.js environment
    if (proxy && isNode && HttpsProxyAgent) {
      fetchOptions.agent = new HttpsProxyAgent(proxy);
    }
    return sendResult(fetch("".concat(getEndpoint(endpoints, path, testnet)).concat(path).concat(makeQueryString(data)), fetchOptions));
  };
};

/**
 * Factory method for partial private calls against the api
 *
 * @param {string} path Endpoint path
 * @param {object} data The payload to be sent
 * @param {string} method HTTB VERB, GET by default
 * @returns {object} The api response
 */
var keyCall = function keyCall(_ref2) {
  var apiKey = _ref2.apiKey,
    pubCall = _ref2.pubCall;
  return function (path, data) {
    var method = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'GET';
    if (!apiKey) {
      throw new Error('You need to pass an API key to make this call.');
    }
    return pubCall(path, data, method, {
      'X-MBX-APIKEY': apiKey
    });
  };
};

/**
 * Factory method for private calls against the api
 *
 * @param {string} path Endpoint path
 * @param {object} data The payload to be sent
 * @param {string} method HTTB VERB, GET by default
 * @param {object} headers
 * @returns {object} The api response
 */
var privateCall = function privateCall(_ref3) {
  var apiKey = _ref3.apiKey,
    apiSecret = _ref3.apiSecret,
    privateKey = _ref3.privateKey,
    proxy = _ref3.proxy,
    endpoints = _ref3.endpoints,
    _ref3$getTime = _ref3.getTime,
    getTime = _ref3$getTime === void 0 ? defaultGetTime : _ref3$getTime,
    pubCall = _ref3.pubCall,
    testnet = _ref3.testnet;
  return function (path) {
    var data = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    var method = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'GET';
    var noData = arguments.length > 3 ? arguments[3] : undefined;
    var noExtra = arguments.length > 4 ? arguments[4] : undefined;
    if (!apiKey || !apiSecret && !privateKey) {
      throw new Error('You need to pass an API key and secret/privateKey to make authenticated calls.');
    }
    return (data && data.useServerTime ? pubCall('/api/v3/time').then(function (r) {
      return r.serverTime;
    }) : Promise.resolve(getTime())).then(function (timestamp) {
      if (data) {
        delete data.useServerTime;
      }
      var queryString = makeQueryString(_objectSpread(_objectSpread({}, data), {}, {
        timestamp: timestamp
      }));
      var dataToSign = queryString.substr(1);

      // Create signature (async in browser, sync in Node.js)
      if (apiSecret) {
        return (0, _signature.createHmacSignature)(dataToSign, apiSecret).then(function (signature) {
          return {
            timestamp: timestamp,
            signature: signature
          };
        });
      } else if (privateKey) {
        var sig = (0, _signature.createAsymmetricSignature)(dataToSign, privateKey);
        // .then(signature => ({
        //     timestamp,
        //     signature,
        // }))
        return {
          timestamp: timestamp,
          signature: sig
        };
      }
    }).then(function (_ref4) {
      var timestamp = _ref4.timestamp,
        signature = _ref4.signature;
      var newData = noExtra ? data : _objectSpread(_objectSpread({}, data), {}, {
        timestamp: timestamp,
        signature: signature
      });
      var fetchOptions = {
        method: method,
        headers: {
          'X-MBX-APIKEY': apiKey
        },
        json: true
      };

      // Only add proxy agent in Node.js environment
      if (proxy && isNode && HttpsProxyAgent) {
        fetchOptions.agent = new HttpsProxyAgent(proxy);
      }
      return sendResult(fetch("".concat(getEndpoint(endpoints, path, testnet)).concat(path).concat(noData ? '' : makeQueryString(newData)), fetchOptions));
    });
  };
};
var candleFields = exports.candleFields = ['openTime', 'open', 'high', 'low', 'close', 'volume', 'closeTime', 'quoteVolume', 'trades', 'baseAssetVolume', 'quoteAssetVolume'];
var deliveryCandleFields = exports.deliveryCandleFields = ['openTime', 'open', 'high', 'low', 'close', 'volume', 'closeTime', 'baseVolume', 'trades', 'quoteAssetVolume', 'baseAssetVolume'];

/**
 * Get candles for a specific pair and interval and convert response
 * to a user friendly collection.
 */
var _candles = function candles(pubCall, payload) {
  var endpoint = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : '/api/v3/klines';
  return checkParams('candles', payload, endpoint.includes('indexPrice') ? ['pair'] : ['symbol']) && pubCall(endpoint, _objectSpread({
    interval: '5m'
  }, payload)).then(function (candles) {
    return candles.map(function (candle) {
      return (0, _lodash.default)(!endpoint.includes('dapi') ? candleFields : deliveryCandleFields, candle);
    });
  });
};
var isContractURL = function isContractURL(path) {
  var isFutures = path.includes('/fapi') || path.includes('/futures');
  var isDelivery = path.includes('/dapi');
  var isPortfolioMargin = path.includes('/papi');
  return isFutures || isDelivery || isPortfolioMargin;
};

/**
 * Create a new order wrapper for market order simplicity
 */
var _order = function order(privCall) {
  var payload = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var url = arguments.length > 2 ? arguments[2] : undefined;
  var newPayload = ['LIMIT', 'STOP_LOSS_LIMIT', 'TAKE_PROFIT_LIMIT'].includes(payload.type) || !payload.type ? _objectSpread({
    timeInForce: 'GTC'
  }, payload) : payload;
  var requires = ['symbol', 'side'];
  if (!(newPayload.type === 'MARKET' && newPayload.quoteOrderQty) && !(newPayload.type === 'STOP_MARKET') && !(newPayload.type === 'TAKE_PROFIT_MARKET') && !(newPayload.type === 'TRAILING_STOP_MARKET')) {
    requires.push('quantity');
  }
  if (newPayload.type === 'TRAILING_STOP_MARKET') {
    requires.push('callbackRate');
  }
  if (!newPayload.newClientOrderId) {
    // check if this is spot or futures/delivery
    var isContract = isContractURL(url);
    if (isContract) {
      newPayload.newClientOrderId = futuresP();
    } else {
      newPayload.newClientOrderId = spotP();
    }
  }
  return checkParams('order', newPayload, requires) && privCall(url, _objectSpread({
    type: 'LIMIT'
  }, newPayload), 'POST');
};
var _orderOco = function orderOco(privCall) {
  var payload = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var url = arguments.length > 2 ? arguments[2] : undefined;
  var newPayload = payload.stopLimitPrice && !payload.stopLimitTimeInForce ? _objectSpread({
    stopLimitTimeInForce: 'GTC'
  }, payload) : payload;
  if (!newPayload.listClientOrderId) {
    newPayload.listClientOrderId = spotP();
  }
  return checkParams('order', newPayload, ['symbol', 'side', 'quantity', 'price', 'stopPrice']) && privCall(url, newPayload, 'POST');
};
var _updateOrder = function updateOrder(privCall) {
  var payload = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var url = arguments.length > 2 ? arguments[2] : undefined;
  var newPayload = _objectSpread({}, payload);
  if (!newPayload.cancelReplaceMode) {
    newPayload.cancelReplaceMode = 'STOP_ON_FAILURE';
  }
  if (!newPayload.timeInForce) {
    newPayload.timeInForce = 'GTC';
  }
  if (!newPayload.newClientOrderId) {
    newPayload.newClientOrderId = spotP();
  }
  return checkParams('updateOrder', newPayload, ['symbol', 'side', 'type']) && privCall(url, newPayload, 'POST');
};

/**
 * Zip asks and bids reponse from order book
 */
var _book = function book(pubCall, payload) {
  var endpoint = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : '/api/v3/depth';
  return checkParams('book', payload, ['symbol']) && pubCall(endpoint, payload).then(function (_ref5) {
    var lastUpdateId = _ref5.lastUpdateId,
      asks = _ref5.asks,
      bids = _ref5.bids;
    return {
      lastUpdateId: lastUpdateId,
      asks: asks.map(function (a) {
        return (0, _lodash.default)(['price', 'quantity'], a);
      }),
      bids: bids.map(function (b) {
        return (0, _lodash.default)(['price', 'quantity'], b);
      })
    };
  });
};
var _aggTrades = function aggTrades(pubCall, payload) {
  var endpoint = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : '/api/v3/aggTrades';
  return checkParams('aggTrades', payload, ['symbol']) && pubCall(endpoint, payload).then(function (trades) {
    return trades.map(function (trade) {
      var transformed = {
        aggId: trade.a,
        symbol: payload.symbol,
        price: trade.p,
        quantity: trade.q,
        firstId: trade.f,
        lastId: trade.l,
        timestamp: trade.T,
        isBuyerMaker: trade.m
      };
      if (trade.M) transformed.wasBestPrice = trade.M;
      return transformed;
    });
  });
};
var _default = exports.default = function _default(opts) {
  var endpoints = {
    base: opts && opts.httpBase,
    futures: opts && opts.httpFutures,
    delivery: opts && opts.httpDelivery,
    portfolioMargin: opts && opts.httpPortfolioMargin
  };
  var pubCall = publicCall(_objectSpread(_objectSpread({}, opts), {}, {
    endpoints: endpoints
  }));
  var deliveryPubCall = publicCall(_objectSpread(_objectSpread({}, opts), {}, {
    endpoints: {
      futures: endpoints.delivery
    }
  }));
  var privCall = privateCall(_objectSpread(_objectSpread({}, opts), {}, {
    endpoints: endpoints,
    pubCall: pubCall
  }));
  var kCall = keyCall(_objectSpread(_objectSpread({}, opts), {}, {
    pubCall: pubCall
  }));
  return {
    // Generic endpoints
    getInfo: function getInfo() {
      return info;
    },
    ping: function ping() {
      return pubCall('/api/v3/ping').then(function () {
        return true;
      });
    },
    time: function time() {
      return pubCall('/api/v3/time').then(function (r) {
        return r.serverTime;
      });
    },
    exchangeInfo: function exchangeInfo(payload) {
      return pubCall('/api/v3/exchangeInfo', payload);
    },
    // Market Data endpoints
    book: function book(payload) {
      return _book(pubCall, payload);
    },
    aggTrades: function aggTrades(payload) {
      return _aggTrades(pubCall, payload);
    },
    candles: function candles(payload) {
      return _candles(pubCall, payload);
    },
    trades: function trades(payload) {
      return checkParams('trades', payload, ['symbol']) && pubCall('/api/v3/trades', payload);
    },
    tradesHistory: function tradesHistory(payload) {
      return checkParams('tradesHitory', payload, ['symbol']) && kCall('/api/v3/historicalTrades', payload);
    },
    dailyStats: function dailyStats(payload) {
      return pubCall('/api/v3/ticker/24hr', payload);
    },
    prices: function prices(payload) {
      return pubCall('/api/v3/ticker/price', payload).then(function (r) {
        return (Array.isArray(r) ? r : [r]).reduce(function (out, cur) {
          return out[cur.symbol] = cur.price, out;
        }, {});
      });
    },
    avgPrice: function avgPrice(payload) {
      return pubCall('/api/v3/avgPrice', payload);
    },
    allBookTickers: function allBookTickers() {
      return pubCall('/api/v3/ticker/bookTicker').then(function (r) {
        return (Array.isArray(r) ? r : [r]).reduce(function (out, cur) {
          return out[cur.symbol] = cur, out;
        }, {});
      });
    },
    // Order endpoints
    order: function order(payload) {
      return _order(privCall, payload, '/api/v3/order');
    },
    updateOrder: function updateOrder(payload) {
      return _updateOrder(privCall, payload, '/api/v3/order/cancelReplace');
    },
    orderOco: function orderOco(payload) {
      return _orderOco(privCall, payload, '/api/v3/order/oco');
    },
    orderTest: function orderTest(payload) {
      return _order(privCall, payload, '/api/v3/order/test');
    },
    getOrder: function getOrder(payload) {
      return privCall('/api/v3/order', payload);
    },
    getOrderOco: function getOrderOco(payload) {
      return privCall('/api/v3/orderList', payload);
    },
    cancelOrder: function cancelOrder(payload) {
      return privCall('/api/v3/order', payload, 'DELETE');
    },
    cancelOrderOco: function cancelOrderOco(payload) {
      return privCall('/api/v3/orderList', payload, 'DELETE');
    },
    cancelOpenOrders: function cancelOpenOrders(payload) {
      return privCall('/api/v3/openOrders', payload, 'DELETE');
    },
    openOrders: function openOrders(payload) {
      return privCall('/api/v3/openOrders', payload);
    },
    allOrders: function allOrders(payload) {
      return privCall('/api/v3/allOrders', payload);
    },
    allOrdersOCO: function allOrdersOCO(payload) {
      return privCall('/api/v3/allOrderList', payload);
    },
    // Account endpoints
    accountInfo: function accountInfo(payload) {
      return privCall('/api/v3/account', payload);
    },
    myTrades: function myTrades(payload) {
      return privCall('/api/v3/myTrades', payload);
    },
    withdraw: function withdraw(payload) {
      return privCall('/sapi/v1/capital/withdraw/apply', payload, 'POST');
    },
    withdrawHistory: function withdrawHistory(payload) {
      return privCall('/sapi/v1/capital/withdraw/history', payload);
    },
    depositHistory: function depositHistory(payload) {
      return privCall('/sapi/v1/capital/deposit/hisrec', payload);
    },
    depositAddress: function depositAddress(payload) {
      return privCall('/sapi/v1/capital/deposit/address', payload);
    },
    tradeFee: function tradeFee(payload) {
      return privCall('/sapi/v1/asset/tradeFee', payload);
    },
    assetDetail: function assetDetail(payload) {
      return privCall('/sapi/v1/asset/assetDetail', payload);
    },
    accountSnapshot: function accountSnapshot(payload) {
      return privCall('/sapi/v1/accountSnapshot', payload);
    },
    universalTransfer: function universalTransfer(payload) {
      return privCall('/sapi/v1/asset/transfer', payload, 'POST');
    },
    universalTransferHistory: function universalTransferHistory(payload) {
      return privCall('/sapi/v1/asset/transfer', payload);
    },
    dustLog: function dustLog(payload) {
      return privCall('/sapi/v1/asset/dribblet', payload);
    },
    dustTransfer: function dustTransfer(payload) {
      return privCall('/sapi/v1/asset/dust', payload, 'POST');
    },
    accountCoins: function accountCoins(payload) {
      return privCall('/sapi/v1/capital/config/getall', payload);
    },
    getBnbBurn: function getBnbBurn(payload) {
      return privCall('/sapi/v1/bnbBurn', payload);
    },
    setBnbBurn: function setBnbBurn(payload) {
      return privCall('/sapi/v1/bnbBurn', payload, 'POST');
    },
    capitalConfigs: function capitalConfigs() {
      return privCall('/sapi/v1/capital/config/getall');
    },
    // User Data Stream endpoints
    getDataStream: function getDataStream() {
      return privCall('/api/v3/userDataStream', null, 'POST', true);
    },
    keepDataStream: function keepDataStream(payload) {
      return privCall('/api/v3/userDataStream', payload, 'PUT', false, true);
    },
    closeDataStream: function closeDataStream(payload) {
      return privCall('/api/v3/userDataStream', payload, 'DELETE', false, true);
    },
    marginGetListenToken: function marginGetListenToken(payload) {
      return privCall('/sapi/v1/userListenToken', payload, 'POST');
    },
    futuresGetDataStream: function futuresGetDataStream() {
      return privCall('/fapi/v1/listenKey', null, 'POST', true);
    },
    futuresKeepDataStream: function futuresKeepDataStream(payload) {
      return privCall('/fapi/v1/listenKey', payload, 'PUT', false, true);
    },
    futuresCloseDataStream: function futuresCloseDataStream(payload) {
      return privCall('/fapi/v1/listenKey', payload, 'DELETE', false, true);
    },
    deliveryGetDataStream: function deliveryGetDataStream() {
      return privCall('/dapi/v1/listenKey', null, 'POST', true);
    },
    deliveryKeepDataStream: function deliveryKeepDataStream(payload) {
      return privCall('/dapi/v1/listenKey', payload, 'PUT', false, true);
    },
    deliveryCloseDataStream: function deliveryCloseDataStream(payload) {
      return privCall('/dapi/v1/listenKey', payload, 'DELETE', false, true);
    },
    // Futures endpoints
    futuresPing: function futuresPing() {
      return pubCall('/fapi/v1/ping').then(function () {
        return true;
      });
    },
    futuresTime: function futuresTime() {
      return pubCall('/fapi/v1/time').then(function (r) {
        return r.serverTime;
      });
    },
    futuresExchangeInfo: function futuresExchangeInfo() {
      return pubCall('/fapi/v1/exchangeInfo');
    },
    futuresBook: function futuresBook(payload) {
      return _book(pubCall, payload, '/fapi/v1/depth');
    },
    futuresAggTrades: function futuresAggTrades(payload) {
      return _aggTrades(pubCall, payload, '/fapi/v1/aggTrades');
    },
    futuresMarkPrice: function futuresMarkPrice(payload) {
      return pubCall('/fapi/v1/premiumIndex', payload);
    },
    futuresAllForceOrders: function futuresAllForceOrders(payload) {
      return pubCall('/fapi/v1/allForceOrders', payload);
    },
    futuresLongShortRatio: function futuresLongShortRatio(payload) {
      return pubCall('/futures/data/globalLongShortAccountRatio', payload);
    },
    futuresCandles: function futuresCandles(payload) {
      return _candles(pubCall, payload, '/fapi/v1/klines');
    },
    futuresMarkPriceCandles: function futuresMarkPriceCandles(payload) {
      return _candles(pubCall, payload, '/fapi/v1/markPriceKlines');
    },
    futuresIndexPriceCandles: function futuresIndexPriceCandles(payload) {
      return _candles(pubCall, payload, '/fapi/v1/indexPriceKlines');
    },
    futuresTrades: function futuresTrades(payload) {
      return checkParams('trades', payload, ['symbol']) && pubCall('/fapi/v1/trades', payload);
    },
    futuresDailyStats: function futuresDailyStats(payload) {
      return pubCall('/fapi/v1/ticker/24hr', payload);
    },
    futuresPrices: function futuresPrices(payload) {
      return pubCall('/fapi/v1/ticker/price', payload).then(function (r) {
        return (Array.isArray(r) ? r : [r]).reduce(function (out, cur) {
          return out[cur.symbol] = cur.price, out;
        }, {});
      });
    },
    futuresAllBookTickers: function futuresAllBookTickers() {
      return pubCall('/fapi/v1/ticker/bookTicker').then(function (r) {
        return (Array.isArray(r) ? r : [r]).reduce(function (out, cur) {
          return out[cur.symbol] = cur, out;
        }, {});
      });
    },
    futuresFundingRate: function futuresFundingRate(payload) {
      return checkParams('fundingRate', payload, ['symbol']) && pubCall('/fapi/v1/fundingRate', payload);
    },
    futuresOrder: function futuresOrder(payload) {
      var _payload$type;
      // Check if this is a conditional order type that should be routed to algo endpoint
      var orderType = payload === null || payload === void 0 || (_payload$type = payload.type) === null || _payload$type === void 0 ? void 0 : _payload$type.toUpperCase();
      var conditionalTypes = ['STOP', 'STOP_MARKET', 'TAKE_PROFIT', 'TAKE_PROFIT_MARKET', 'TRAILING_STOP_MARKET'];
      if (orderType && conditionalTypes.includes(orderType)) {
        // Route to algo order endpoint
        var algoPayload = _objectSpread({}, payload);
        if (!algoPayload.clientAlgoId) {
          algoPayload.clientAlgoId = futuresP();
        }
        delete algoPayload.newClientOrderId;
        algoPayload.algoType = 'CONDITIONAL';
        if (algoPayload.stopPrice && !algoPayload.triggerPrice) {
          algoPayload.triggerPrice = algoPayload.stopPrice;
          delete algoPayload.stopPrice;
        }
        return privCall('/fapi/v1/algoOrder', algoPayload, 'POST');
      }
      // Use regular order endpoint
      return _order(privCall, payload, '/fapi/v1/order');
    },
    futuresUpdateOrder: function futuresUpdateOrder(payload) {
      if (payload && 'conditional' in payload) {
        // for now it is not supported
        // const payloadCopy = { ...payload }
        // delete payloadCopy.conditional
        // return privCall('/fapi/v1/algoOrder', payloadCopy, 'PUT')
      }
      return privCall('/fapi/v1/order', payload, 'PUT');
    },
    futuresBatchOrders: function futuresBatchOrders(payload) {
      return privCall('/fapi/v1/batchOrders', payload, 'POST');
    },
    futuresGetOrder: function futuresGetOrder(payload) {
      // Check if this is a request for a conditional/algo order
      var isConditional = payload === null || payload === void 0 ? void 0 : payload.conditional;
      var hasAlgoId = (payload === null || payload === void 0 ? void 0 : payload.algoId) || (payload === null || payload === void 0 ? void 0 : payload.clientAlgoId);
      var payloadCopy = payload;
      if (payload && 'conditional' in payload) {
        payloadCopy = _objectSpread({}, payload);
        delete payloadCopy.conditional;
      }
      if (isConditional || hasAlgoId) {
        return privCall('/fapi/v1/algoOrder', payloadCopy);
      }
      return privCall('/fapi/v1/order', payloadCopy);
    },
    futuresCancelOrder: function futuresCancelOrder(payload) {
      // Check if this is a request for a conditional/algo order
      var isConditional = payload === null || payload === void 0 ? void 0 : payload.conditional;
      var hasAlgoId = (payload === null || payload === void 0 ? void 0 : payload.algoId) || (payload === null || payload === void 0 ? void 0 : payload.clientAlgoId);
      var payloadCopy = payload;
      if (payload && 'conditional' in payload) {
        payloadCopy = _objectSpread({}, payload);
        delete payloadCopy.conditional;
      }
      if (isConditional || hasAlgoId) {
        return privCall('/fapi/v1/algoOrder', payloadCopy, 'DELETE');
      }
      return privCall('/fapi/v1/order', payloadCopy, 'DELETE');
    },
    futuresCancelAllOpenOrders: function futuresCancelAllOpenOrders(payload) {
      var isConditional = payload === null || payload === void 0 ? void 0 : payload.conditional;
      var payloadCopy = payload;
      if (payload && 'conditional' in payload) {
        payloadCopy = _objectSpread({}, payload);
        delete payloadCopy.conditional;
      }
      if (isConditional) {
        return privCall('/fapi/v1/algoOpenOrders', payloadCopy, 'DELETE');
      }
      return privCall('/fapi/v1/allOpenOrders', payloadCopy, 'DELETE');
    },
    futuresCancelBatchOrders: function futuresCancelBatchOrders(payload) {
      return privCall('/fapi/v1/batchOrders', payload, 'DELETE');
    },
    futuresOpenOrders: function futuresOpenOrders(payload) {
      var isConditional = payload === null || payload === void 0 ? void 0 : payload.conditional;
      var payloadCopy = payload;
      if (payload && 'conditional' in payload) {
        payloadCopy = _objectSpread({}, payload);
        delete payloadCopy.conditional;
      }
      if (isConditional) {
        return privCall('/fapi/v1/openAlgoOrders', payloadCopy);
      }
      return privCall('/fapi/v1/openOrders', payloadCopy);
    },
    futuresAllOrders: function futuresAllOrders(payload) {
      var isConditional = payload === null || payload === void 0 ? void 0 : payload.conditional;
      var payloadCopy = payload;
      if (payload && 'conditional' in payload) {
        payloadCopy = _objectSpread({}, payload);
        delete payloadCopy.conditional;
      }
      if (isConditional) {
        return privCall('/fapi/v1/allAlgoOrders', payloadCopy);
      }
      return privCall('/fapi/v1/allOrders', payloadCopy);
    },
    futuresPositionRisk: function futuresPositionRisk(payload) {
      return privCall('/fapi/v2/positionRisk', payload);
    },
    futuresLeverageBracket: function futuresLeverageBracket(payload) {
      return privCall('/fapi/v1/leverageBracket', payload);
    },
    futuresAccountBalance: function futuresAccountBalance(payload) {
      return privCall('/fapi/v2/balance', payload);
    },
    futuresAccountInfo: function futuresAccountInfo(payload) {
      return privCall('/fapi/v2/account', payload);
    },
    futuresUserTrades: function futuresUserTrades(payload) {
      return privCall('/fapi/v1/userTrades', payload);
    },
    futuresPositionMode: function futuresPositionMode(payload) {
      return privCall('/fapi/v1/positionSide/dual', payload);
    },
    futuresPositionModeChange: function futuresPositionModeChange(payload) {
      return privCall('/fapi/v1/positionSide/dual', payload, 'POST');
    },
    futuresLeverage: function futuresLeverage(payload) {
      return privCall('/fapi/v1/leverage', payload, 'POST');
    },
    futuresMarginType: function futuresMarginType(payload) {
      return privCall('/fapi/v1/marginType', payload, 'POST');
    },
    futuresPositionMargin: function futuresPositionMargin(payload) {
      return privCall('/fapi/v1/positionMargin', payload, 'POST');
    },
    futuresMarginHistory: function futuresMarginHistory(payload) {
      return privCall('/fapi/v1/positionMargin/history', payload);
    },
    futuresIncome: function futuresIncome(payload) {
      return privCall('/fapi/v1/income', payload);
    },
    getMultiAssetsMargin: function getMultiAssetsMargin(payload) {
      return privCall('/fapi/v1/multiAssetsMargin', payload);
    },
    setMultiAssetsMargin: function setMultiAssetsMargin(payload) {
      return privCall('/fapi/v1/multiAssetsMargin', payload, 'POST');
    },
    futuresRpiDepth: function futuresRpiDepth(payload) {
      return _book(pubCall, payload, '/fapi/v1/rpiDepth');
    },
    futuresSymbolAdlRisk: function futuresSymbolAdlRisk(payload) {
      return pubCall('/fapi/v1/symbolAdlRisk', payload);
    },
    futuresCommissionRate: function futuresCommissionRate(payload) {
      return privCall('/fapi/v1/commissionRate', payload);
    },
    // Algo Orders (Conditional Orders)
    futuresCreateAlgoOrder: function futuresCreateAlgoOrder(payload) {
      if (!payload.clientAlgoId) {
        payload.clientAlgoId = futuresP();
      }
      if (!payload.algoType) {
        payload.algoType = 'CONDITIONAL';
      }
      return privCall('/fapi/v1/algoOrder', payload, 'POST');
    },
    futuresCancelAlgoOrder: function futuresCancelAlgoOrder(payload) {
      return privCall('/fapi/v1/algoOrder', payload, 'DELETE');
    },
    futuresCancelAllAlgoOpenOrders: function futuresCancelAllAlgoOpenOrders(payload) {
      return privCall('/fapi/v1/algoOpenOrders', payload, 'DELETE');
    },
    futuresGetAlgoOrder: function futuresGetAlgoOrder(payload) {
      return privCall('/fapi/v1/algoOrder', payload);
    },
    futuresGetOpenAlgoOrders: function futuresGetOpenAlgoOrders(payload) {
      return privCall('/fapi/v1/openAlgoOrders', payload);
    },
    futuresGetAllAlgoOrders: function futuresGetAllAlgoOrders(payload) {
      return privCall('/fapi/v1/allAlgoOrders', payload);
    },
    // Delivery endpoints
    deliveryPing: function deliveryPing() {
      return pubCall('/dapi/v1/ping').then(function () {
        return true;
      });
    },
    deliveryTime: function deliveryTime() {
      return pubCall('/dapi/v1/time').then(function (r) {
        return r.serverTime;
      });
    },
    deliveryExchangeInfo: function deliveryExchangeInfo() {
      return pubCall('/dapi/v1/exchangeInfo');
    },
    deliveryBook: function deliveryBook(payload) {
      return _book(pubCall, payload, '/dapi/v1/depth');
    },
    deliveryAggTrades: function deliveryAggTrades(payload) {
      return _aggTrades(pubCall, payload, '/dapi/v1/aggTrades');
    },
    deliveryMarkPrice: function deliveryMarkPrice(payload) {
      return pubCall('/dapi/v1/premiumIndex', payload);
    },
    deliveryAllForceOrders: function deliveryAllForceOrders(payload) {
      return pubCall('/dapi/v1/allForceOrders', payload);
    },
    deliveryLongShortRatio: function deliveryLongShortRatio(payload) {
      return deliveryPubCall('/futures/data/globalLongShortAccountRatio', payload);
    },
    deliveryCandles: function deliveryCandles(payload) {
      return _candles(pubCall, payload, '/dapi/v1/klines');
    },
    deliveryMarkPriceCandles: function deliveryMarkPriceCandles(payload) {
      return _candles(pubCall, payload, '/dapi/v1/markPriceKlines');
    },
    deliveryIndexPriceCandles: function deliveryIndexPriceCandles(payload) {
      return _candles(pubCall, payload, '/dapi/v1/indexPriceKlines');
    },
    deliveryTrades: function deliveryTrades(payload) {
      return checkParams('trades', payload, ['symbol']) && pubCall('/dapi/v1/trades', payload);
    },
    deliveryDailyStats: function deliveryDailyStats(payload) {
      return pubCall('/dapi/v1/ticker/24hr', payload);
    },
    deliveryPrices: function deliveryPrices() {
      return pubCall('/dapi/v1/ticker/price').then(function (r) {
        return (Array.isArray(r) ? r : [r]).reduce(function (out, cur) {
          return out[cur.symbol] = cur.price, out;
        }, {});
      });
    },
    deliveryAllBookTickers: function deliveryAllBookTickers() {
      return pubCall('/dapi/v1/ticker/bookTicker').then(function (r) {
        return (Array.isArray(r) ? r : [r]).reduce(function (out, cur) {
          return out[cur.symbol] = cur, out;
        }, {});
      });
    },
    deliveryFundingRate: function deliveryFundingRate(payload) {
      return checkParams('fundingRate', payload, ['symbol']) && pubCall('/dapi/v1/fundingRate', payload);
    },
    deliveryOrder: function deliveryOrder(payload) {
      return _order(privCall, payload, '/dapi/v1/order');
    },
    deliveryBatchOrders: function deliveryBatchOrders(payload) {
      return privCall('/dapi/v1/batchOrders', payload, 'POST');
    },
    deliveryGetOrder: function deliveryGetOrder(payload) {
      return privCall('/dapi/v1/order', payload);
    },
    deliveryCancelOrder: function deliveryCancelOrder(payload) {
      return privCall('/dapi/v1/order', payload, 'DELETE');
    },
    deliveryCancelAllOpenOrders: function deliveryCancelAllOpenOrders(payload) {
      return privCall('/dapi/v1/allOpenOrders', payload, 'DELETE');
    },
    deliveryCancelBatchOrders: function deliveryCancelBatchOrders(payload) {
      return privCall('/dapi/v1/batchOrders', payload, 'DELETE');
    },
    deliveryOpenOrders: function deliveryOpenOrders(payload) {
      return privCall('/dapi/v1/openOrders', payload);
    },
    deliveryAllOrders: function deliveryAllOrders(payload) {
      return privCall('/dapi/v1/allOrders', payload);
    },
    deliveryPositionRisk: function deliveryPositionRisk(payload) {
      return privCall('/dapi/v1/positionRisk', payload);
    },
    deliveryLeverageBracket: function deliveryLeverageBracket(payload) {
      return privCall('/dapi/v1/leverageBracket', payload);
    },
    deliveryAccountBalance: function deliveryAccountBalance(payload) {
      return privCall('/dapi/v1/balance', payload);
    },
    deliveryAccountInfo: function deliveryAccountInfo(payload) {
      return privCall('/dapi/v1/account', payload);
    },
    deliveryUserTrades: function deliveryUserTrades(payload) {
      return privCall('/dapi/v1/userTrades', payload);
    },
    deliveryPositionMode: function deliveryPositionMode(payload) {
      return privCall('/dapi/v1/positionSide/dual', payload);
    },
    deliveryPositionModeChange: function deliveryPositionModeChange(payload) {
      return privCall('/dapi/v1/positionSide/dual', payload, 'POST');
    },
    deliveryLeverage: function deliveryLeverage(payload) {
      return privCall('/dapi/v1/leverage', payload, 'POST');
    },
    deliveryMarginType: function deliveryMarginType(payload) {
      return privCall('/dapi/v1/marginType', payload, 'POST');
    },
    deliveryPositionMargin: function deliveryPositionMargin(payload) {
      return privCall('/dapi/v1/positionMargin', payload, 'POST');
    },
    deliveryMarginHistory: function deliveryMarginHistory(payload) {
      return privCall('/dapi/v1/positionMargin/history', payload);
    },
    deliveryIncome: function deliveryIncome(payload) {
      return privCall('/dapi/v1/income', payload);
    },
    // PAPI endpoints
    papiPing: function papiPing() {
      return privCall('/papi/v1/ping');
    },
    papiAccount: function papiAccount() {
      return privCall('/papi/v1/account');
    },
    papiBalance: function papiBalance(payload) {
      return privCall('/papi/v1/balance', payload);
    },
    papiUmOrder: function papiUmOrder(payload) {
      return privCall('/papi/v1/um/order', payload);
    },
    papiUmConditionalOrder: function papiUmConditionalOrder(payload) {
      return privCall('/papi/v1/um/conditional/order', payload, 'POST');
    },
    papiCmOrder: function papiCmOrder(payload) {
      return privCall('/papi/v1/cm/order', payload, 'POST');
    },
    papiCmConditionalOrder: function papiCmConditionalOrder(payload) {
      return privCall('/papi/v1/cm/conditional/order', payload, 'POST');
    },
    papiMarginOrder: function papiMarginOrder(payload) {
      return privCall('/papi/v1/margin/order', payload, 'POST');
    },
    papiMarginLoan: function papiMarginLoan(payload) {
      return privCall('/papi/v1/marginLoan', payload, 'POST');
    },
    papiRepayLoan: function papiRepayLoan(payload) {
      return privCall('/papi/v1/repayLoan', payload, 'POST');
    },
    papiMarginOrderOco: function papiMarginOrderOco(payload) {
      return privCall('/papi/v1/margin/order/oco', payload, 'POST');
    },
    papiUmCancelOrder: function papiUmCancelOrder(payload) {
      return privCall('/papi/v1/um/order', payload, 'DELETE');
    },
    papiUmCancelAllOpenOrders: function papiUmCancelAllOpenOrders(payload) {
      return privCall('/papi/v1/um/allOpenOrders', payload, 'DELETE');
    },
    papiUmCancelConditionalOrder: function papiUmCancelConditionalOrder(payload) {
      return privCall('/papi/v1/um/conditional/order', payload, 'DELETE');
    },
    papiUmCancelConditionalAllOpenOrders: function papiUmCancelConditionalAllOpenOrders(payload) {
      return privCall('/papi/v1/um/conditional/allOpenOrders', payload, 'DELETE');
    },
    papiCmCancelOrder: function papiCmCancelOrder(payload) {
      return privCall('/papi/v1/cm/order', payload, 'DELETE');
    },
    papiCmCancelAllOpenOrders: function papiCmCancelAllOpenOrders(payload) {
      return privCall('/papi/v1/cm/allOpenOrders', payload, 'DELETE');
    },
    papiCmCancelConditionalOrder: function papiCmCancelConditionalOrder(payload) {
      return privCall('/papi/v1/cm/conditional/order', payload, 'DELETE');
    },
    papiCmCancelConditionalAllOpenOrders: function papiCmCancelConditionalAllOpenOrders(payload) {
      return privCall('/papi/v1/cm/conditional/allOpenOrders', payload, 'DELETE');
    },
    papiMarginCancelOrder: function papiMarginCancelOrder(payload) {
      return privCall('/papi/v1/margin/order', payload, 'DELETE');
    },
    papiMarginCancelOrderList: function papiMarginCancelOrderList(payload) {
      return privCall('/papi/v1/margin/orderList', payload, 'DELETE');
    },
    papiMarginCancelAllOpenOrders: function papiMarginCancelAllOpenOrders(payload) {
      return privCall('/papi/v1/margin/allOpenOrders', payload, 'DELETE');
    },
    papiUmUpdateOrder: function papiUmUpdateOrder(payload) {
      return privCall('/papi/v1/um/order', payload, 'PUT');
    },
    papiCmUpdateOrder: function papiCmUpdateOrder(payload) {
      return privCall('/papi/v1/cm/order', payload, 'PUT');
    },
    papiUmGetOrder: function papiUmGetOrder(payload) {
      return privCall('/papi/v1/um/order', payload);
    },
    papiUmGetAllOrders: function papiUmGetAllOrders(payload) {
      return privCall('/papi/v1/um/allOrders', payload);
    },
    papiUmGetOpenOrder: function papiUmGetOpenOrder(payload) {
      return privCall('/papi/v1/um/openOrder', payload);
    },
    papiUmGetOpenOrders: function papiUmGetOpenOrders(payload) {
      return privCall('/papi/v1/um/openOrders', payload);
    },
    papiUmGetConditionalAllOrders: function papiUmGetConditionalAllOrders(payload) {
      return privCall('/papi/v1/um/conditional/allOrders', payload);
    },
    papiUmGetConditionalOpenOrders: function papiUmGetConditionalOpenOrders(payload) {
      return privCall('/papi/v1/um/conditional/openOrders', payload);
    },
    papiUmGetConditionalOpenOrder: function papiUmGetConditionalOpenOrder(payload) {
      return privCall('/papi/v1/um/conditional/openOrder', payload);
    },
    papiUmGetConditionalOrderHistory: function papiUmGetConditionalOrderHistory(payload) {
      return privCall('/papi/v1/um/conditional/orderHistory', payload);
    },
    papiCmGetOrder: function papiCmGetOrder(payload) {
      return privCall('/papi/v1/cm/order', payload);
    },
    papiCmGetAllOrders: function papiCmGetAllOrders(payload) {
      return privCall('/papi/v1/cm/allOrders', payload);
    },
    papiCmGetOpenOrder: function papiCmGetOpenOrder(payload) {
      return privCall('/papi/v1/cm/openOrder', payload);
    },
    papiCmGetOpenOrders: function papiCmGetOpenOrders(payload) {
      return privCall('/papi/v1/cm/openOrders', payload);
    },
    papiCmGetConditionalOpenOrders: function papiCmGetConditionalOpenOrders(payload) {
      return privCall('/papi/v1/cm/conditional/openOrders', payload);
    },
    papiCmGetConditionalOpenOrder: function papiCmGetConditionalOpenOrder(payload) {
      return privCall('/papi/v1/cm/conditional/openOrder', payload);
    },
    papiCmGetConditionalAllOrders: function papiCmGetConditionalAllOrders(payload) {
      return privCall('/papi/v1/cm/conditional/allOrders', payload);
    },
    papiCmGetConditionalOrderHistory: function papiCmGetConditionalOrderHistory(payload) {
      return privCall('/papi/v1/cm/conditional/orderHistory', payload);
    },
    papiUmGetForceOrders: function papiUmGetForceOrders(payload) {
      return privCall('/papi/v1/um/forceOrders', payload);
    },
    papiCmGetForceOrders: function papiCmGetForceOrders(payload) {
      return privCall('/papi/v1/cm/forceOrders', payload);
    },
    papiUmGetOrderAmendment: function papiUmGetOrderAmendment(payload) {
      return privCall('/papi/v1/um/orderAmendment', payload);
    },
    papiCmGetOrderAmendment: function papiCmGetOrderAmendment(payload) {
      return privCall('/papi/v1/cm/orderAmendment', payload);
    },
    papiMarginGetForceOrders: function papiMarginGetForceOrders(payload) {
      return privCall('/papi/v1/margin/forceOrders', payload);
    },
    papiUmGetUserTrades: function papiUmGetUserTrades(payload) {
      return privCall('/papi/v1/um/userTrades', payload);
    },
    papiCmGetUserTrades: function papiCmGetUserTrades(payload) {
      return privCall('/papi/v1/cm/userTrades', payload);
    },
    papiUmGetAdlQuantile: function papiUmGetAdlQuantile(payload) {
      return privCall('/papi/v1/um/adlQuantile', payload);
    },
    papiCmGetAdlQuantile: function papiCmGetAdlQuantile(payload) {
      return privCall('/papi/v1/cm/adlQuantile', payload);
    },
    papiUmFeeBurn: function papiUmFeeBurn(payload) {
      return privCall('/papi/v1/um/feeBurn', payload, 'POST');
    },
    papiUmGetFeeBurn: function papiUmGetFeeBurn(payload) {
      return privCall('/papi/v1/um/feeBurn', payload);
    },
    papiMarginGetOrder: function papiMarginGetOrder(payload) {
      return privCall('/papi/v1/margin/order', payload);
    },
    papiMarginGetOpenOrders: function papiMarginGetOpenOrders(payload) {
      return privCall('/papi/v1/margin/openOrders', payload);
    },
    papiMarginGetAllOrders: function papiMarginGetAllOrders(payload) {
      return privCall('/papi/v1/margin/allOrders', payload);
    },
    papiMarginGetOrderList: function papiMarginGetOrderList(payload) {
      return privCall('/papi/v1/margin/orderList', payload);
    },
    papiMarginGetAllOrderList: function papiMarginGetAllOrderList(payload) {
      return privCall('/papi/v1/margin/allOrderList', payload);
    },
    papiMarginGetOpenOrderList: function papiMarginGetOpenOrderList(payload) {
      return privCall('/papi/v1/margin/openOrderList', payload);
    },
    papiMarginGetMyTrades: function papiMarginGetMyTrades(payload) {
      return privCall('/papi/v1/margin/myTrades', payload);
    },
    papiMarginRepayDebt: function papiMarginRepayDebt(payload) {
      return privCall('/papi/v1/margin/repay-debt', payload, 'POST');
    },
    // Margin endpoints
    marginAllOrders: function marginAllOrders(payload) {
      return privCall('/sapi/v1/margin/allOrders', payload);
    },
    marginOrder: function marginOrder(payload) {
      return _order(privCall, payload, '/sapi/v1/margin/order');
    },
    marginOrderOco: function marginOrderOco(payload) {
      return _orderOco(privCall, payload, '/sapi/v1/margin/order/oco');
    },
    marginGetOrder: function marginGetOrder(payload) {
      return privCall('/sapi/v1/margin/order', payload);
    },
    marginGetOrderOco: function marginGetOrderOco(payload) {
      return privCall('/sapi/v1/margin/orderList', payload);
    },
    marginCancelOrder: function marginCancelOrder(payload) {
      return privCall('/sapi/v1/margin/order', payload, 'DELETE');
    },
    marginOpenOrders: function marginOpenOrders(payload) {
      return privCall('/sapi/v1/margin/openOrders', payload);
    },
    marginCancelOpenOrders: function marginCancelOpenOrders(payload) {
      return privCall('/sapi/v1/margin/openOrders', payload, 'DELETE');
    },
    marginAccountInfo: function marginAccountInfo(payload) {
      return privCall('/sapi/v1/margin/account', payload);
    },
    marginMyTrades: function marginMyTrades(payload) {
      return privCall('/sapi/v1/margin/myTrades', payload);
    },
    marginRepay: function marginRepay(payload) {
      return privCall('/sapi/v1/margin/repay', payload, 'POST');
    },
    marginLoan: function marginLoan(payload) {
      return privCall('/sapi/v1/margin/loan', payload, 'POST');
    },
    marginCapitalFlow: function marginCapitalFlow(payload) {
      return privCall('/sapi/v1/margin/capital-flow', payload);
    },
    marginInterestHistory: function marginInterestHistory(payload) {
      return privCall('/sapi/v1/margin/interestHistory', payload);
    },
    marginIsolatedAccount: function marginIsolatedAccount(payload) {
      return privCall('/sapi/v1/margin/isolated/account', payload);
    },
    marginMaxBorrow: function marginMaxBorrow(payload) {
      return privCall('/sapi/v1/margin/maxBorrowable', payload);
    },
    marginCreateIsolated: function marginCreateIsolated(payload) {
      return privCall('/sapi/v1/margin/isolated/create', payload, 'POST');
    },
    marginIsolatedTransfer: function marginIsolatedTransfer(payload) {
      return privCall('/sapi/v1/margin/isolated/transfer', payload, 'POST');
    },
    marginIsolatedTransferHistory: function marginIsolatedTransferHistory(payload) {
      return privCall('/sapi/v1/margin/isolated/transfer', payload);
    },
    disableMarginAccount: function disableMarginAccount(payload) {
      return privCall('/sapi/v1/margin/isolated/account', payload, 'DELETE');
    },
    enableMarginAccount: function enableMarginAccount(payload) {
      return privCall('/sapi/v1/margin/isolated/account', payload, 'POST');
    },
    isolatedMarginAllPairs: function isolatedMarginAllPairs(payload) {
      return privCall('/sapi/v1/margin/isolated/allPairs', payload);
    },
    isolatedMarginAccount: function isolatedMarginAccount(payload) {
      return privCall('/sapi/v1/margin/isolated/account', payload);
    },
    marginAccount: function marginAccount() {
      return privCall('/sapi/v1/margin/account');
    },
    // Portfolio Margin endpoints
    portfolioMarginAccountInfo: function portfolioMarginAccountInfo() {
      return privCall('/sapi/v1/portfolio/account');
    },
    portfolioMarginCollateralRate: function portfolioMarginCollateralRate() {
      return privCall('/sapi/v1/portfolio/collateralRate');
    },
    portfolioMarginLoan: function portfolioMarginLoan(payload) {
      return privCall('/sapi/v1/portfolio/pmLoan', payload);
    },
    portfolioMarginLoanRepay: function portfolioMarginLoanRepay(payload) {
      return privCall('/sapi/v1/portfolio/repay', payload, 'POST');
    },
    portfolioMarginInterestHistory: function portfolioMarginInterestHistory(payload) {
      return privCall('/sapi/v1/portfolio/interest-history', payload);
    },
    // Savings endpoints
    savingsAccount: function savingsAccount(payload) {
      return privCall('/sapi/v1/lending/union/account', payload);
    },
    savingsPurchase: function savingsPurchase(payload) {
      return privCall('/sapi/v1/lending/union/purchase', payload, 'POST');
    },
    savingsRedeem: function savingsRedeem(payload) {
      return privCall('/sapi/v1/lending/union/redeem', payload, 'POST');
    },
    fundingWallet: function fundingWallet(payload) {
      return privCall('/sapi/v1/asset/get-funding-asset', payload, 'POST');
    },
    convertTradeFlow: function convertTradeFlow(payload) {
      return privCall('/sapi/v1/convert/tradeFlow', payload);
    },
    rebateTaxQuery: function rebateTaxQuery() {
      return privCall('/sapi/v1/rebate/taxQuery');
    },
    payTradeHistory: function payTradeHistory(payload) {
      return privCall('/sapi/v1/pay/transactions', payload);
    },
    apiRestrictions: function apiRestrictions(payload) {
      return privCall('/sapi/v1/account/apiRestrictions', payload);
    },
    // Mining endpoints
    miningHashrateResaleRequest: function miningHashrateResaleRequest(payload) {
      return privCall('/sapi/v1/mining/hash-transfer/config', payload, 'POST');
    },
    miningHashrateResaleCancel: function miningHashrateResaleCancel(payload) {
      return privCall('/sapi/v1/mining/hash-transfer/config/cancel', payload, 'POST');
    },
    miningStatistics: function miningStatistics(payload) {
      return privCall('/sapi/v1/mining/statistics/user/status', payload);
    },
    // Utility endpoints
    privateRequest: function privateRequest(method, url, payload) {
      return privCall(url, payload, method);
    },
    publicRequest: function publicRequest(method, url, payload) {
      return pubCall(url, payload, method);
    }
  };
};
