// binance-api-node/dist/websocket.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.userEventHandler = exports.keepStreamAlive = exports.default = void 0;
var _lodash = _interopRequireDefault(require("lodash.zipobject"));
var _jsonBigint = _interopRequireDefault(require("json-bigint"));
var _httpClient = _interopRequireDefault(require("./http-client"));
var _openWebsocket = _interopRequireDefault(require("./open-websocket"));
var _signature = require("./signature");
var _excluded = ["e"];
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _toConsumableArray(r) { return _arrayWithoutHoles(r) || _iterableToArray(r) || _unsupportedIterableToArray(r) || _nonIterableSpread(); }
function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _iterableToArray(r) { if ("undefined" != typeof Symbol && null != r[Symbol.iterator] || null != r["@@iterator"]) return Array.from(r); }
function _arrayWithoutHoles(r) { if (Array.isArray(r)) return _arrayLikeToArray(r); }
function _objectWithoutProperties(e, t) { if (null == e) return {}; var o, r, i = _objectWithoutPropertiesLoose(e, t); if (Object.getOwnPropertySymbols) { var n = Object.getOwnPropertySymbols(e); for (r = 0; r < n.length; r++) o = n[r], -1 === t.indexOf(o) && {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]); } return i; }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (-1 !== e.indexOf(n)) continue; t[n] = r[n]; } return t; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _slicedToArray(r, e) { return _arrayWithHoles(r) || _iterableToArrayLimit(r, e) || _unsupportedIterableToArray(r, e) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t.return && (u = t.return(), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(r) { if (Array.isArray(r)) return r; }
var endpoints = {
  base: 'wss://stream.binance.com:9443/ws',
  futures: 'wss://fstream.binance.com/market/ws',
  futuresPublic: 'wss://fstream.binance.com/public/ws',
  futuresMarket: 'wss://fstream.binance.com/market/ws',
  futuresPrivate: 'wss://fstream.binance.com/private/ws',
  delivery: 'wss://dstream.binance.com/ws'
};
var wsOptions = {};
function openWebSocket(url) {
  return (0, _openWebsocket.default)(url, wsOptions);
}
var depthTransform = function depthTransform(m) {
  return {
    eventType: m.e,
    eventTime: m.E,
    symbol: m.s,
    firstUpdateId: m.U,
    finalUpdateId: m.u,
    bidDepth: m.b.map(function (b) {
      return (0, _lodash.default)(['price', 'quantity'], b);
    }),
    askDepth: m.a.map(function (a) {
      return (0, _lodash.default)(['price', 'quantity'], a);
    })
  };
};
var futuresDepthTransform = function futuresDepthTransform(m) {
  return {
    eventType: m.e,
    eventTime: m.E,
    transactionTime: m.T,
    symbol: m.s,
    firstUpdateId: m.U,
    finalUpdateId: m.u,
    prevFinalUpdateId: m.pu,
    bidDepth: m.b.map(function (b) {
      return (0, _lodash.default)(['price', 'quantity'], b);
    }),
    askDepth: m.a.map(function (a) {
      return (0, _lodash.default)(['price', 'quantity'], a);
    })
  };
};
var deliveryDepthTransform = function deliveryDepthTransform(m) {
  return {
    eventType: m.e,
    eventTime: m.E,
    transactionTime: m.T,
    symbol: m.s,
    pair: m.ps,
    firstUpdateId: m.U,
    finalUpdateId: m.u,
    prevFinalUpdateId: m.pu,
    bidDepth: m.b.map(function (b) {
      return (0, _lodash.default)(['price', 'quantity'], b);
    }),
    askDepth: m.a.map(function (a) {
      return (0, _lodash.default)(['price', 'quantity'], a);
    })
  };
};
var depth = function depth(payload, cb) {
  var transform = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
  var variator = arguments.length > 3 ? arguments[3] : undefined;
  var cache = (Array.isArray(payload) ? payload : [payload]).map(function (symbol) {
    var _symbol$toLowerCase$s = symbol.toLowerCase().split('@'),
      _symbol$toLowerCase$s2 = _slicedToArray(_symbol$toLowerCase$s, 2),
      symbolName = _symbol$toLowerCase$s2[0],
      updateSpeed = _symbol$toLowerCase$s2[1];
    var w = openWebSocket("".concat(variator === 'futures' ? endpoints.futuresPublic : variator ? endpoints[variator] : endpoints.base, "/").concat(symbolName, "@depth").concat(updateSpeed ? "@".concat(updateSpeed) : ''));
    w.onmessage = function (msg) {
      var obj = _jsonBigint.default.parse(msg.data);
      cb(transform ? variator === 'futures' ? futuresDepthTransform(obj) : variator === 'delivery' ? deliveryDepthTransform(obj) : depthTransform(obj) : obj);
    };
    return w;
  });
  return function (options) {
    return cache.forEach(function (w) {
      return w.close(1000, 'Close handle was called', _objectSpread({
        keepClosed: true
      }, options));
    });
  };
};
var futuresRpiDepth = function futuresRpiDepth(payload, cb) {
  var transform = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
  var cache = (Array.isArray(payload) ? payload : [payload]).map(function (symbol) {
    var symbolName = symbol.toLowerCase();
    var w = openWebSocket("".concat(endpoints.futuresPublic, "/").concat(symbolName, "@rpiDepth@500ms"));
    w.onmessage = function (msg) {
      var obj = _jsonBigint.default.parse(msg.data);
      cb(transform ? futuresDepthTransform(obj) : obj);
    };
    return w;
  });
  return function (options) {
    return cache.forEach(function (w) {
      return w.close(1000, 'Close handle was called', _objectSpread({
        keepClosed: true
      }, options));
    });
  };
};
var partialDepthTransform = function partialDepthTransform(symbol, level, m) {
  return {
    symbol: symbol,
    level: level,
    lastUpdateId: m.lastUpdateId,
    bids: m.bids.map(function (b) {
      return (0, _lodash.default)(['price', 'quantity'], b);
    }),
    asks: m.asks.map(function (a) {
      return (0, _lodash.default)(['price', 'quantity'], a);
    })
  };
};
var futuresPartDepthTransform = function futuresPartDepthTransform(level, m) {
  return {
    level: level,
    eventType: m.e,
    eventTime: m.E,
    transactionTime: m.T,
    symbol: m.s,
    firstUpdateId: m.U,
    finalUpdateId: m.u,
    prevFinalUpdateId: m.pu,
    bidDepth: m.b.map(function (b) {
      return (0, _lodash.default)(['price', 'quantity'], b);
    }),
    askDepth: m.a.map(function (a) {
      return (0, _lodash.default)(['price', 'quantity'], a);
    })
  };
};
var deliveryPartDepthTransform = function deliveryPartDepthTransform(level, m) {
  return {
    level: level,
    eventType: m.e,
    eventTime: m.E,
    transactionTime: m.T,
    symbol: m.s,
    pair: m.ps,
    firstUpdateId: m.U,
    finalUpdateId: m.u,
    prevFinalUpdateId: m.pu,
    bidDepth: m.b.map(function (b) {
      return (0, _lodash.default)(['price', 'quantity'], b);
    }),
    askDepth: m.a.map(function (a) {
      return (0, _lodash.default)(['price', 'quantity'], a);
    })
  };
};
var partialDepth = function partialDepth(payload, cb) {
  var transform = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
  var variator = arguments.length > 3 ? arguments[3] : undefined;
  var cache = (Array.isArray(payload) ? payload : [payload]).map(function (_ref) {
    var symbol = _ref.symbol,
      level = _ref.level;
    var _symbol$toLowerCase$s3 = symbol.toLowerCase().split('@'),
      _symbol$toLowerCase$s4 = _slicedToArray(_symbol$toLowerCase$s3, 2),
      symbolName = _symbol$toLowerCase$s4[0],
      updateSpeed = _symbol$toLowerCase$s4[1];
    var w = openWebSocket("".concat(variator === 'futures' ? endpoints.futuresPublic : variator ? endpoints[variator] : endpoints.base, "/").concat(symbolName, "@depth").concat(level).concat(updateSpeed ? "@".concat(updateSpeed) : ''));
    w.onmessage = function (msg) {
      var obj = _jsonBigint.default.parse(msg.data);
      cb(transform ? variator === 'futures' ? futuresPartDepthTransform(level, obj) : variator === 'delivery' ? deliveryPartDepthTransform(level, obj) : partialDepthTransform(symbol, level, obj) : obj);
    };
    return w;
  });
  return function (options) {
    return cache.forEach(function (w) {
      return w.close(1000, 'Close handle was called', _objectSpread({
        keepClosed: true
      }, options));
    });
  };
};
var candleTransform = function candleTransform(m) {
  return {
    startTime: m.t,
    closeTime: m.T,
    firstTradeId: m.f,
    lastTradeId: m.L,
    open: m.o,
    high: m.h,
    low: m.l,
    close: m.c,
    volume: m.v,
    trades: m.n,
    interval: m.i,
    isFinal: m.x,
    quoteVolume: m.q,
    buyVolume: m.V,
    quoteBuyVolume: m.Q
  };
};
var deliveryCandleTransform = function deliveryCandleTransform(m) {
  return {
    startTime: m.t,
    closeTime: m.T,
    firstTradeId: m.f,
    lastTradeId: m.L,
    open: m.o,
    high: m.h,
    low: m.l,
    close: m.c,
    volume: m.v,
    trades: m.n,
    interval: m.i,
    isFinal: m.x,
    baseVolume: m.q,
    buyVolume: m.V,
    baseBuyVolume: m.Q
  };
};
var candles = function candles(payload, interval, cb) {
  var transform = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : true;
  var variator = arguments.length > 4 ? arguments[4] : undefined;
  if (!interval || !cb) {
    throw new Error('Please pass a symbol, interval and callback.');
  }
  var cache = (Array.isArray(payload) ? payload : [payload]).map(function (symbol) {
    var w = openWebSocket("".concat(variator === 'futures' ? endpoints.futuresMarket : variator === 'delivery' ? endpoints.delivery : endpoints.base, "/").concat(symbol.toLowerCase(), "@kline_").concat(interval));
    w.onmessage = function (msg) {
      var obj = _jsonBigint.default.parse(msg.data);
      var eventType = obj.e,
        eventTime = obj.E,
        symbol = obj.s,
        tick = obj.k;
      cb(transform ? _objectSpread({
        eventType: eventType,
        eventTime: eventTime,
        symbol: symbol
      }, variator === 'delivery' ? deliveryCandleTransform(tick) : candleTransform(tick)) : obj);
    };
    return w;
  });
  return function (options) {
    return cache.forEach(function (w) {
      return w.close(1000, 'Close handle was called', _objectSpread({
        keepClosed: true
      }, options));
    });
  };
};
var bookTickerTransform = function bookTickerTransform(m) {
  return {
    updateId: m.u,
    symbol: m.s,
    bestBid: m.b,
    bestBidQnt: m.B,
    bestAsk: m.a,
    bestAskQnt: m.A
  };
};
var miniTickerTransform = function miniTickerTransform(m) {
  return {
    eventType: m.e,
    eventTime: m.E,
    symbol: m.s,
    curDayClose: m.c,
    open: m.o,
    high: m.h,
    low: m.l,
    volume: m.v,
    volumeQuote: m.q
  };
};
var deliveryMiniTickerTransform = function deliveryMiniTickerTransform(m) {
  return {
    eventType: m.e,
    eventTime: m.E,
    symbol: m.s,
    pair: m.ps,
    curDayClose: m.c,
    open: m.o,
    high: m.h,
    low: m.l,
    volume: m.v,
    volumeBase: m.q
  };
};
var tickerTransform = function tickerTransform(m) {
  return {
    eventType: m.e,
    eventTime: m.E,
    symbol: m.s,
    priceChange: m.p,
    priceChangePercent: m.P,
    weightedAvg: m.w,
    prevDayClose: m.x,
    curDayClose: m.c,
    closeTradeQuantity: m.Q,
    bestBid: m.b,
    bestBidQnt: m.B,
    bestAsk: m.a,
    bestAskQnt: m.A,
    open: m.o,
    high: m.h,
    low: m.l,
    volume: m.v,
    volumeQuote: m.q,
    openTime: m.O,
    closeTime: m.C,
    firstTradeId: m.F,
    lastTradeId: m.L,
    totalTrades: m.n
  };
};
var futuresTickerTransform = function futuresTickerTransform(m) {
  return {
    eventType: m.e,
    eventTime: m.E,
    symbol: m.s,
    priceChange: m.p,
    priceChangePercent: m.P,
    weightedAvg: m.w,
    curDayClose: m.c,
    closeTradeQuantity: m.Q,
    open: m.o,
    high: m.h,
    low: m.l,
    volume: m.v,
    volumeQuote: m.q,
    openTime: m.O,
    closeTime: m.C,
    firstTradeId: m.F,
    lastTradeId: m.L,
    totalTrades: m.n
  };
};
var deliveryTickerTransform = function deliveryTickerTransform(m) {
  return {
    eventType: m.e,
    eventTime: m.E,
    symbol: m.s,
    pair: m.ps,
    priceChange: m.p,
    priceChangePercent: m.P,
    weightedAvg: m.w,
    curDayClose: m.c,
    closeTradeQuantity: m.Q,
    open: m.o,
    high: m.h,
    low: m.l,
    volume: m.v,
    volumeBase: m.q,
    openTime: m.O,
    closeTime: m.C,
    firstTradeId: m.F,
    lastTradeId: m.L,
    totalTrades: m.n
  };
};
var bookTicker = function bookTicker(payload, cb) {
  var transform = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
  var cache = (Array.isArray(payload) ? payload : [payload]).map(function (symbol) {
    var w = openWebSocket("".concat(endpoints.base, "/").concat(symbol.toLowerCase(), "@bookTicker"));
    w.onmessage = function (msg) {
      var obj = _jsonBigint.default.parse(msg.data);
      cb(transform ? bookTickerTransform(obj) : obj);
    };
    return w;
  });
  return function (options) {
    return cache.forEach(function (w) {
      return w.close(1000, 'Close handle was called', _objectSpread({
        keepClosed: true
      }, options));
    });
  };
};
var ticker = function ticker(payload, cb) {
  var transform = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
  var variator = arguments.length > 3 ? arguments[3] : undefined;
  var cache = (Array.isArray(payload) ? payload : [payload]).map(function (symbol) {
    var w = openWebSocket("".concat(variator === 'futures' ? endpoints.futuresMarket : variator === 'delivery' ? endpoints.delivery : endpoints.base, "/").concat(symbol.toLowerCase(), "@ticker"));
    w.onmessage = function (msg) {
      var obj = _jsonBigint.default.parse(msg.data);
      cb(transform ? variator === 'futures' ? futuresTickerTransform(obj) : variator === 'delivery' ? deliveryTickerTransform(obj) : tickerTransform(obj) : obj);
    };
    return w;
  });
  return function (options) {
    return cache.forEach(function (w) {
      return w.close(1000, 'Close handle was called', _objectSpread({
        keepClosed: true
      }, options));
    });
  };
};
var allTickers = function allTickers(cb) {
  var transform = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
  var variator = arguments.length > 2 ? arguments[2] : undefined;
  var w = new openWebSocket("".concat(variator === 'futures' ? endpoints.futuresMarket : variator === 'delivery' ? endpoints.delivery : endpoints.base, "/!miniTicker@arr"));
  w.onmessage = function (msg) {
    var arr = _jsonBigint.default.parse(msg.data);
    cb(transform ? variator === 'futures' ? arr.map(function (m) {
      return futuresTickerTransform(m);
    }) : variator === 'delivery' ? arr.map(function (m) {
      return deliveryTickerTransform(m);
    }) : arr.map(function (m) {
      return tickerTransform(m);
    }) : arr);
  };
  return function (options) {
    return w.close(1000, 'Close handle was called', _objectSpread({
      keepClosed: true
    }, options));
  };
};
var allTickersDeprecated = function allTickersDeprecated(cb) {
  var transform = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
  var variator = arguments.length > 2 ? arguments[2] : undefined;
  var w = new openWebSocket("".concat(variator === 'futures' ? endpoints.futures : variator === 'delivery' ? endpoints.delivery : endpoints.base, "/!ticker@arr"));
  w.onmessage = function (msg) {
    var arr = _jsonBigint.default.parse(msg.data);
    cb(transform ? variator === 'futures' ? arr.map(function (m) {
      return futuresTickerTransform(m);
    }) : variator === 'delivery' ? arr.map(function (m) {
      return deliveryTickerTransform(m);
    }) : arr.map(function (m) {
      return tickerTransform(m);
    }) : arr);
  };
  return function (options) {
    return w.close(1000, 'Close handle was called', _objectSpread({
      keepClosed: true
    }, options));
  };
};
var miniTicker = function miniTicker(payload, cb) {
  var transform = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
  var variator = arguments.length > 3 ? arguments[3] : undefined;
  var cache = (Array.isArray(payload) ? payload : [payload]).map(function (symbol) {
    var w = openWebSocket("".concat(endpoints.base, "/").concat(symbol.toLowerCase(), "@miniTicker"));
    w.onmessage = function (msg) {
      var obj = _jsonBigint.default.parse(msg.data);
      cb(transform ? variator === 'delivery' ? deliveryMiniTickerTransform(obj) : miniTickerTransform(obj) : obj);
    };
    return w;
  });
  return function (options) {
    return cache.forEach(function (w) {
      return w.close(1000, 'Close handle was called', _objectSpread({
        keepClosed: true
      }, options));
    });
  };
};
var allMiniTickers = function allMiniTickers(cb) {
  var transform = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
  var variator = arguments.length > 2 ? arguments[2] : undefined;
  var w = openWebSocket("".concat(endpoints.base, "/!miniTicker@arr"));
  w.onmessage = function (msg) {
    var arr = _jsonBigint.default.parse(msg.data);
    cb(transform ? arr.map(variator === 'delivery' ? deliveryMiniTickerTransform : miniTickerTransform) : arr);
  };
  return function (options) {
    return function (w) {
      return w.close(1000, 'Close handle was called', _objectSpread({
        keepClosed: true
      }, options));
    };
  };
};
var customSubStream = function customSubStream(payload, cb, variator) {
  var cache = (Array.isArray(payload) ? payload : [payload]).map(function (sub) {
    var w = openWebSocket("".concat(variator === 'futures' ? endpoints.futures : variator === 'delivery' ? endpoints.delivery : endpoints.base, "/").concat(sub));
    w.onmessage = function (msg) {
      var data = _jsonBigint.default.parse(msg.data);
      cb(data);
    };
    return w;
  });
  return function (options) {
    return cache.forEach(function (w) {
      return w.close(1000, 'Close handle was called', _objectSpread({
        keepClosed: true
      }, options));
    });
  };
};
var aggTradesTransform = function aggTradesTransform(m) {
  return {
    eventType: m.e,
    eventTime: m.E,
    timestamp: m.T,
    symbol: m.s,
    price: m.p,
    quantity: m.q,
    isBuyerMaker: m.m,
    wasBestPrice: m.M,
    aggId: m.a,
    firstId: m.f,
    lastId: m.l
  };
};
var futuresAggTradesTransform = function futuresAggTradesTransform(m) {
  return {
    eventType: m.e,
    eventTime: m.E,
    symbol: m.s,
    aggId: m.a,
    price: m.p,
    quantity: m.q,
    firstId: m.f,
    lastId: m.l,
    timestamp: m.T,
    isBuyerMaker: m.m
  };
};
var aggTrades = function aggTrades(payload, cb) {
  var transform = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
  var variator = arguments.length > 3 ? arguments[3] : undefined;
  var cache = (Array.isArray(payload) ? payload : [payload]).map(function (symbol) {
    var w = openWebSocket("".concat(variator === 'futures' ? endpoints.futuresMarket : variator === 'delivery' ? endpoints.delivery : endpoints.base, "/").concat(symbol.toLowerCase(), "@aggTrade"));
    w.onmessage = function (msg) {
      var obj = _jsonBigint.default.parse(msg.data);
      cb(transform ? variator === 'futures' || variator === 'delivery' ? futuresAggTradesTransform(obj) : aggTradesTransform(obj) : obj);
    };
    return w;
  });
  return function (options) {
    return cache.forEach(function (w) {
      return w.close(1000, 'Close handle was called', _objectSpread({
        keepClosed: true
      }, options));
    });
  };
};
var futuresLiqsTransform = function futuresLiqsTransform(m) {
  return {
    symbol: m.s,
    price: m.p,
    origQty: m.q,
    lastFilledQty: m.l,
    accumulatedQty: m.z,
    averagePrice: m.ap,
    status: m.X,
    timeInForce: m.f,
    type: m.o,
    side: m.S,
    time: m.T
  };
};
var futuresLiquidations = function futuresLiquidations(payload, cb) {
  var transform = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
  var cache = (Array.isArray(payload) ? payload : [payload]).map(function (symbol) {
    var w = openWebSocket("".concat(endpoints.futuresMarket, "/").concat(symbol.toLowerCase(), "@forceOrder"));
    w.onmessage = function (msg) {
      var obj = _jsonBigint.default.parse(msg.data);
      cb(transform ? futuresLiqsTransform(obj.o) : obj);
    };
    return w;
  });
  return function (options) {
    return cache.forEach(function (w) {
      return w.close(1000, 'Close handle was called', _objectSpread({
        keepClosed: true
      }, options));
    });
  };
};
var futuresAllLiquidations = function futuresAllLiquidations(cb) {
  var transform = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
  var w = new openWebSocket("".concat(endpoints.futuresMarket, "/!forceOrder@arr"));
  w.onmessage = function (msg) {
    var obj = _jsonBigint.default.parse(msg.data);
    cb(transform ? futuresLiqsTransform(obj.o) : obj);
  };
  return function (options) {
    return w.close(1000, 'Close handle was called', _objectSpread({
      keepClosed: true
    }, options));
  };
};
var futuresBookTicker = function futuresBookTicker(payload, cb) {
  var transform = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
  var cache = (Array.isArray(payload) ? payload : [payload]).map(function (symbol) {
    var w = openWebSocket("".concat(endpoints.futuresPublic, "/").concat(symbol.toLowerCase(), "@bookTicker"));
    w.onmessage = function (msg) {
      var obj = _jsonBigint.default.parse(msg.data);
      cb(transform ? bookTickerTransform(obj) : obj);
    };
    return w;
  });
  return function (options) {
    return cache.forEach(function (w) {
      return w.close(1000, 'Close handle was called', _objectSpread({
        keepClosed: true
      }, options));
    });
  };
};
var futuresAllBookTickers = function futuresAllBookTickers(cb) {
  var transform = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
  var w = openWebSocket("".concat(endpoints.futuresPublic, "/!bookTicker"));
  w.onmessage = function (msg) {
    var obj = _jsonBigint.default.parse(msg.data);
    cb(transform ? bookTickerTransform(obj) : obj);
  };
  return function (options) {
    return w.close(1000, 'Close handle was called', _objectSpread({
      keepClosed: true
    }, options));
  };
};
var futuresMarkPriceTransform = function futuresMarkPriceTransform(m) {
  return {
    eventType: m.e,
    eventTime: m.E,
    symbol: m.s,
    markPrice: m.p,
    indexPrice: m.i,
    settlePrice: m.P,
    fundingRate: m.r,
    nextFundingRate: m.T
  };
};
var futuresMarkPrice = function futuresMarkPrice(payload, cb) {
  var transform = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
  var cache = (Array.isArray(payload) ? payload : [payload]).map(function (input) {
    var symbol = _typeof(input) === 'object' ? input.symbol : input;
    var updateSpeed = _typeof(input) === 'object' ? input.updateSpeed : undefined;
    var stream = updateSpeed === '1s' ? "".concat(symbol.toLowerCase(), "@markPrice@1s") : "".concat(symbol.toLowerCase(), "@markPrice");
    var w = openWebSocket("".concat(endpoints.futuresMarket, "/").concat(stream));
    w.onmessage = function (msg) {
      var obj = _jsonBigint.default.parse(msg.data);
      cb(transform ? futuresMarkPriceTransform(obj) : obj);
    };
    return w;
  });
  return function (options) {
    return cache.forEach(function (w) {
      return w.close(1000, 'Close handle was called', _objectSpread({
        keepClosed: true
      }, options));
    });
  };
};
var futuresContinuousCandles = function futuresContinuousCandles(payload, interval, cb) {
  var transform = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : true;
  if (!interval || !cb) {
    throw new Error('Please pass a pair, contractType, interval and callback.');
  }
  var pair = payload.pair.toLowerCase();
  var contractType = payload.contractType.toLowerCase();
  var w = openWebSocket("".concat(endpoints.futuresMarket, "/").concat(pair, "_").concat(contractType, "@continuousKline_").concat(interval));
  w.onmessage = function (msg) {
    var obj = _jsonBigint.default.parse(msg.data);
    var eventType = obj.e,
      eventTime = obj.E,
      pairSymbol = obj.ps,
      contType = obj.ct,
      tick = obj.k;
    cb(transform ? _objectSpread({
      eventType: eventType,
      eventTime: eventTime,
      pair: pairSymbol,
      contractType: contType
    }, candleTransform(tick)) : obj);
  };
  return function (options) {
    return w.close(1000, 'Close handle was called', _objectSpread({
      keepClosed: true
    }, options));
  };
};
var futuresCompositeIndex = function futuresCompositeIndex(payload, cb) {
  var transform = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
  var cache = (Array.isArray(payload) ? payload : [payload]).map(function (symbol) {
    var w = openWebSocket("".concat(endpoints.futuresMarket, "/").concat(symbol.toLowerCase(), "@compositeIndex"));
    w.onmessage = function (msg) {
      var obj = _jsonBigint.default.parse(msg.data);
      cb(transform ? {
        eventType: obj.e,
        eventTime: obj.E,
        symbol: obj.s,
        price: obj.p,
        composition: obj.c ? obj.c.map(function (c) {
          return {
            baseAsset: c.b,
            quoteAsset: c.q,
            weightInQuantity: c.w,
            weightInPercentage: c.W,
            indexPrice: c.i
          };
        }) : []
      } : obj);
    };
    return w;
  });
  return function (options) {
    return cache.forEach(function (w) {
      return w.close(1000, 'Close handle was called', _objectSpread({
        keepClosed: true
      }, options));
    });
  };
};
var futuresContractInfo = function futuresContractInfo(cb) {
  var transform = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
  var w = openWebSocket("".concat(endpoints.futuresMarket, "/!contractInfo"));
  w.onmessage = function (msg) {
    var obj = _jsonBigint.default.parse(msg.data);
    cb(transform ? {
      eventType: obj.e,
      eventTime: obj.E,
      symbol: obj.s,
      pair: obj.ps,
      contractType: obj.ct,
      deliveryDate: obj.dt,
      onboardDate: obj.ot,
      contractStatus: obj.cs,
      brackets: obj.bks ? obj.bks.map(function (b) {
        return {
          notionalBracket: b.bs,
          floorNotional: b.bnf,
          capNotional: b.bnc,
          maintenanceRatio: b.mmr,
          auxiliaryNumber: b.cf,
          minLeverage: b.mi,
          maxLeverage: b.ma
        };
      }) : []
    } : obj);
  };
  return function (options) {
    return w.close(1000, 'Close handle was called', _objectSpread({
      keepClosed: true
    }, options));
  };
};
var futuresAssetIndexTransform = function futuresAssetIndexTransform(m) {
  return {
    eventType: m.e,
    eventTime: m.E,
    symbol: m.s,
    index: m.i,
    bidBuffer: m.b,
    askBuffer: m.a,
    bidRate: m.B,
    askRate: m.A,
    autoExchangeBidBuffer: m.q,
    autoExchangeAskBuffer: m.Q,
    autoExchangeBidRate: m.g,
    autoExchangeAskRate: m.G
  };
};
var futuresAssetIndex = function futuresAssetIndex(payload, cb) {
  var transform = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
  var cache = (Array.isArray(payload) ? payload : [payload]).map(function (symbol) {
    var w = openWebSocket("".concat(endpoints.futuresMarket, "/").concat(symbol.toLowerCase(), "@assetIndex"));
    w.onmessage = function (msg) {
      var obj = _jsonBigint.default.parse(msg.data);
      cb(transform ? futuresAssetIndexTransform(obj) : obj);
    };
    return w;
  });
  return function (options) {
    return cache.forEach(function (w) {
      return w.close(1000, 'Close handle was called', _objectSpread({
        keepClosed: true
      }, options));
    });
  };
};
var futuresAllAssetIndex = function futuresAllAssetIndex(cb) {
  var transform = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
  var w = openWebSocket("".concat(endpoints.futuresMarket, "/!assetIndex@arr"));
  w.onmessage = function (msg) {
    var arr = _jsonBigint.default.parse(msg.data);
    cb(transform ? arr.map(futuresAssetIndexTransform) : arr);
  };
  return function (options) {
    return w.close(1000, 'Close handle was called', _objectSpread({
      keepClosed: true
    }, options));
  };
};
var tradesTransform = function tradesTransform(m) {
  return {
    eventType: m.e,
    eventTime: m.E,
    tradeTime: m.T,
    symbol: m.s,
    price: m.p,
    quantity: m.q,
    isBuyerMaker: m.m,
    maker: m.M,
    tradeId: m.t,
    buyerOrderId: m.b,
    sellerOrderId: m.a
  };
};
var trades = function trades(payload, cb) {
  var transform = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
  var cache = (Array.isArray(payload) ? payload : [payload]).map(function (symbol) {
    var w = openWebSocket("".concat(endpoints.base, "/").concat(symbol.toLowerCase(), "@trade"));
    w.onmessage = function (msg) {
      var obj = _jsonBigint.default.parse(msg.data);
      cb(transform ? tradesTransform(obj) : obj);
    };
    return w;
  });
  return function (options) {
    return cache.forEach(function (w) {
      return w.close(1000, 'Close handle was called', _objectSpread({
        keepClosed: true
      }, options));
    });
  };
};
var userTransforms = {
  // https://github.com/binance-exchange/binance-official-api-docs/blob/master/user-data-stream.md#balance-update
  balanceUpdate: function balanceUpdate(m) {
    return {
      asset: m.a,
      balanceDelta: m.d,
      clearTime: m.T,
      eventTime: m.E,
      eventType: 'balanceUpdate'
    };
  },
  // https://github.com/binance-exchange/binance-official-api-docs/blob/master/user-data-stream.md#account-update
  outboundAccountInfo: function outboundAccountInfo(m) {
    return {
      eventType: 'account',
      eventTime: m.E,
      makerCommissionRate: m.m,
      takerCommissionRate: m.t,
      buyerCommissionRate: m.b,
      sellerCommissionRate: m.s,
      canTrade: m.T,
      canWithdraw: m.W,
      canDeposit: m.D,
      lastAccountUpdate: m.u,
      balances: m.B.reduce(function (out, cur) {
        out[cur.a] = {
          available: cur.f,
          locked: cur.l
        };
        return out;
      }, {})
    };
  },
  // https://github.com/binance-exchange/binance-official-api-docs/blob/master/user-data-stream.md#account-update
  outboundAccountPosition: function outboundAccountPosition(m) {
    return {
      balances: m.B.map(function (_ref2) {
        var a = _ref2.a,
          f = _ref2.f,
          l = _ref2.l;
        return {
          asset: a,
          free: f,
          locked: l
        };
      }),
      eventTime: m.E,
      eventType: 'outboundAccountPosition',
      lastAccountUpdate: m.u
    };
  },
  // https://github.com/binance-exchange/binance-official-api-docs/blob/master/user-data-stream.md#order-update
  executionReport: function executionReport(m) {
    return {
      eventType: 'executionReport',
      eventTime: m.E,
      symbol: m.s,
      newClientOrderId: m.c,
      originalClientOrderId: m.C,
      side: m.S,
      orderType: m.o,
      timeInForce: m.f,
      quantity: m.q,
      price: m.p,
      executionType: m.x,
      stopPrice: m.P,
      trailingDelta: m.d,
      icebergQuantity: m.F,
      orderStatus: m.X,
      orderRejectReason: m.r,
      orderId: m.i,
      orderTime: m.T,
      lastTradeQuantity: m.l,
      totalTradeQuantity: m.z,
      priceLastTrade: m.L,
      commission: m.n,
      commissionAsset: m.N,
      tradeId: m.t,
      isOrderWorking: m.w,
      isBuyerMaker: m.m,
      creationTime: m.O,
      totalQuoteTradeQuantity: m.Z,
      orderListId: m.g,
      quoteOrderQuantity: m.Q,
      lastQuoteTransacted: m.Y,
      trailingTime: m.D
    };
  },
  listStatus: function listStatus(m) {
    return {
      eventType: 'listStatus',
      eventTime: m.E,
      symbol: m.s,
      orderListId: m.g,
      contingencyType: m.c,
      listStatusType: m.l,
      listOrderStatus: m.L,
      listRejectReason: m.r,
      listClientOrderId: m.C,
      transactionTime: m.T,
      orders: m.O.map(function (o) {
        return {
          symbol: o.s,
          orderId: o.i,
          clientOrderId: o.c
        };
      })
    };
  }
};
var futuresUserTransforms = {
  // https://binance-docs.github.io/apidocs/futures/en/#close-user-data-stream-user_stream
  listenKeyExpired: function USER_DATA_STREAM_EXPIRED(m) {
    return {
      eventTime: m.E,
      eventType: 'USER_DATA_STREAM_EXPIRED'
    };
  },
  // https://binance-docs.github.io/apidocs/futures/en/#event-margin-call
  MARGIN_CALL: function MARGIN_CALL(m) {
    return {
      eventTime: m.E,
      crossWalletBalance: m.cw,
      eventType: 'MARGIN_CALL',
      positions: m.p.map(function (cur) {
        return {
          symbol: cur.s,
          positionSide: cur.ps,
          positionAmount: cur.pa,
          marginType: cur.mt,
          isolatedWallet: cur.iw,
          markPrice: cur.mp,
          unrealizedPnL: cur.up,
          maintenanceMarginRequired: cur.mm
        };
      })
    };
  },
  // https://binance-docs.github.io/apidocs/futures/en/#event-balance-and-position-update
  ACCOUNT_UPDATE: function ACCOUNT_UPDATE(m) {
    return {
      eventTime: m.E,
      transactionTime: m.T,
      eventType: 'ACCOUNT_UPDATE',
      eventReasonType: m.a.m,
      balances: m.a.B.map(function (b) {
        return {
          asset: b.a,
          walletBalance: b.wb,
          crossWalletBalance: b.cw,
          balanceChange: b.bc
        };
      }),
      positions: m.a.P.map(function (p) {
        return {
          symbol: p.s,
          positionAmount: p.pa,
          entryPrice: p.ep,
          accumulatedRealized: p.cr,
          unrealizedPnL: p.up,
          marginType: p.mt,
          isolatedWallet: p.iw,
          positionSide: p.ps
        };
      })
    };
  },
  // https://binance-docs.github.io/apidocs/futures/en/#event-order-update
  ORDER_TRADE_UPDATE: function ORDER_TRADE_UPDATE(m) {
    return {
      eventType: 'ORDER_TRADE_UPDATE',
      eventTime: m.E,
      transactionTime: m.T,
      symbol: m.o.s,
      clientOrderId: m.o.c,
      side: m.o.S,
      orderType: m.o.o,
      timeInForce: m.o.f,
      quantity: m.o.q,
      price: m.o.p,
      averagePrice: m.o.ap,
      stopPrice: m.o.sp,
      executionType: m.o.x,
      orderStatus: m.o.X,
      orderId: m.o.i,
      lastTradeQuantity: m.o.l,
      totalTradeQuantity: m.o.z,
      priceLastTrade: m.o.L,
      commissionAsset: m.o.N,
      commission: m.o.n,
      orderTime: m.o.T,
      tradeId: m.o.t,
      bidsNotional: m.o.b,
      asksNotional: m.o.a,
      isMaker: m.o.m,
      isReduceOnly: m.o.R,
      workingType: m.o.wt,
      originalOrderType: m.o.ot,
      positionSide: m.o.ps,
      closePosition: m.o.cp,
      activationPrice: m.o.AP,
      callbackRate: m.o.cr,
      realizedProfit: m.o.rp
    };
  },
  // https://binance-docs.github.io/apidocs/futures/en/#event-account-configuration-update-previous-leverage-update
  ACCOUNT_CONFIG_UPDATE: function ACCOUNT_CONFIG_UPDATE(m) {
    return _objectSpread({
      eventType: 'ACCOUNT_CONFIG_UPDATE',
      eventTime: m.E,
      transactionTime: m.T,
      type: m.ac ? 'ACCOUNT_CONFIG' : 'MULTI_ASSETS'
    }, m.ac ? {
      symbol: m.ac.s,
      leverage: m.ac.l
    } : {
      multiAssets: m.ai.j
    });
  }
};
var userEventHandler = exports.userEventHandler = function userEventHandler(cb) {
  var transform = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
  var variator = arguments.length > 2 ? arguments[2] : undefined;
  return function (msg) {
    var _JSONbig$parse = _jsonBigint.default.parse(msg.data),
      type = _JSONbig$parse.e,
      rest = _objectWithoutProperties(_JSONbig$parse, _excluded);
    cb(variator === 'futures' || variator === 'delivery' ? transform && futuresUserTransforms[type] ? futuresUserTransforms[type](rest) : _objectSpread({
      type: type
    }, rest) : transform && userTransforms[type] ? userTransforms[type](rest) : _objectSpread({
      type: type
    }, rest));
  };
};
var userOpenHandler = function userOpenHandler(cb) {
  var transform = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
  return function () {
    cb(_defineProperty({}, transform ? 'eventType' : 'type', 'open'));
  };
};
var userErrorHandler = function userErrorHandler(cb) {
  var transform = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
  return function (error) {
    cb(_defineProperty(_defineProperty({}, transform ? 'eventType' : 'type', 'error'), "error", error));
  };
};
var STREAM_METHODS = ['get', 'keep', 'close'];
var capitalize = function capitalize(str, check) {
  return check ? "".concat(str[0].toUpperCase()).concat(str.slice(1)) : str;
};
var getStreamMethods = function getStreamMethods(opts) {
  var variator = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';
  var methods = (0, _httpClient.default)(opts);
  return STREAM_METHODS.reduce(function (acc, key) {
    return [].concat(_toConsumableArray(acc), [methods["".concat(variator).concat(capitalize("".concat(key, "DataStream"), !!variator))]]);
  }, []);
};
var keepStreamAlive = exports.keepStreamAlive = function keepStreamAlive(method, listenKey) {
  return method({
    listenKey: listenKey
  });
};
var userWebSocketApi = function userWebSocketApi(opts) {
  return function (cb, transform) {
    var isDemo = opts.testnet && !(opts.httpBase && opts.httpBase.includes('testnet'));
    var isTestnet = !isDemo && opts.httpBase && opts.httpBase.includes('testnet');
    var wsApiUrl = opts.wsApi || (isDemo ? 'wss://demo-ws-api.binance.com/ws-api/v3' : isTestnet ? opts.wsApiTestnet || 'wss://ws-api.testnet.binance.vision/ws-api/v3' : 'wss://ws-api.binance.com:443/ws-api/v3');
    var requestId = 1;
    var errorHandler = userErrorHandler(cb, transform);
    var w = openWebSocket(wsApiUrl);
    var sendSubscribe = function sendSubscribe() {
      var timestamp = opts.getTime ? opts.getTime() : Date.now();
      var paramsStr = "apiKey=".concat(opts.apiKey, "&timestamp=").concat(timestamp);
      var doSend = function doSend(signature) {
        w.send(_jsonBigint.default.stringify({
          id: requestId++,
          method: 'userDataStream.subscribe.signature',
          params: {
            apiKey: opts.apiKey,
            timestamp: timestamp,
            signature: signature
          }
        }));
      };
      if (opts.apiSecret) {
        (0, _signature.createHmacSignature)(paramsStr, opts.apiSecret).then(doSend);
      } else if (opts.privateKey) {
        doSend((0, _signature.createAsymmetricSignature)(paramsStr, opts.privateKey));
      }
    };
    return new Promise(function (resolve, reject) {
      var resolved = false;
      w.onopen = function () {
        sendSubscribe();
        if (opts.emitSocketOpens) {
          userOpenHandler(cb, transform)();
        }
      };
      w.onmessage = function (msg) {
        var data = _jsonBigint.default.parse(msg.data);

        // Control response (subscription/unsubscription)
        if ('id' in data) {
          if (data.error) {
            var err = new Error(data.error.msg || 'WebSocket API error');
            err.code = data.error.code;
            if (!resolved) {
              resolved = true;
              reject(err);
            } else if (opts.emitStreamErrors) {
              errorHandler(err);
            }
          } else if (!resolved) {
            resolved = true;
            resolve(function (options) {
              try {
                w.send(_jsonBigint.default.stringify({
                  id: requestId++,
                  method: 'userDataStream.unsubscribe'
                }));
              } catch (e) {
                // Ignore send errors during close
              }
              w.close(1000, 'Close handle was called', _objectSpread({
                keepClosed: true
              }, options));
            });
          }
          return;
        }

        // User data event - unwrap if in wrapped format
        if (cb) {
          var eventData = data;
          if (data.event && _typeof(data.event) === 'object') {
            eventData = data.event;
          }
          if (eventData.e) {
            userEventHandler(cb, transform)({
              data: _jsonBigint.default.stringify(eventData)
            });
          }
        }
      };
      w.onerror = function (event) {
        var error = event.error || event.message || new Error('WebSocket error');
        if (opts.emitSocketErrors) {
          errorHandler(typeof error === 'string' ? new Error(error) : error);
        }
      };
    });
  };
};
var marginUserWebSocketApi = function marginUserWebSocketApi(opts) {
  return function (cb, transform) {
    var marginOpts = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
    var isTestnet = opts.testnet || opts.httpBase && opts.httpBase.includes('testnet');
    var wsApiUrl = isTestnet ? opts.wsApiTestnet || 'wss://ws-api.testnet.binance.vision/ws-api/v3' : opts.wsApi || 'wss://ws-api.binance.com:443/ws-api/v3';
    var methods = (0, _httpClient.default)(opts);
    var requestId = 1;
    var renewalTimeout = null;
    var w = null;
    var keepClosed = false;
    var errorHandler = userErrorHandler(cb, transform);
    var RENEWAL_BUFFER_MS = 5 * 60 * 1000;
    var cleanup = function cleanup() {
      var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var internal = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
      if (!internal) keepClosed = true;
      if (renewalTimeout) {
        clearTimeout(renewalTimeout);
        renewalTimeout = null;
      }
      if (w) {
        try {
          w.send(_jsonBigint.default.stringify({
            id: requestId++,
            method: 'userDataStream.unsubscribe'
          }));
        } catch (e) {
          // Ignore send errors during close
        }
        w.close(1000, 'Close handle was called', _objectSpread({
          keepClosed: !internal
        }, options));
        w = null;
      }
    };
    var scheduleRenewal = function scheduleRenewal(expirationTime) {
      if (renewalTimeout) clearTimeout(renewalTimeout);
      var delay = Math.max(expirationTime - Date.now() - RENEWAL_BUFFER_MS, 60000);
      renewalTimeout = setTimeout(function () {
        return _renewToken();
      }, delay);
    };
    var _renewToken = function renewToken() {
      if (keepClosed || !w) return;
      methods.marginGetListenToken(marginOpts).then(function (_ref3) {
        var token = _ref3.token,
          expirationTime = _ref3.expirationTime;
        if (keepClosed || !w) return;
        w.send(_jsonBigint.default.stringify({
          id: requestId++,
          method: 'userDataStream.subscribe.listenToken',
          params: {
            listenToken: token
          }
        }));
        scheduleRenewal(expirationTime);
      }).catch(function (err) {
        if (opts.emitStreamErrors) errorHandler(err);
        if (!keepClosed) {
          renewalTimeout = setTimeout(function () {
            return _renewToken();
          }, 30e3);
        }
      });
    };
    var _makeStream = function makeStream(isReconnecting) {
      if (keepClosed) return Promise.resolve();
      return methods.marginGetListenToken(marginOpts).then(function (_ref4) {
        var token = _ref4.token,
          expirationTime = _ref4.expirationTime;
        if (keepClosed) return;
        w = openWebSocket(wsApiUrl);
        return new Promise(function (resolve, reject) {
          var resolved = false;
          w.onopen = function () {
            w.send(_jsonBigint.default.stringify({
              id: requestId++,
              method: 'userDataStream.subscribe.listenToken',
              params: {
                listenToken: token
              }
            }));
            if (opts.emitSocketOpens) {
              userOpenHandler(cb, transform)();
            }
          };
          w.onmessage = function (msg) {
            var data = _jsonBigint.default.parse(msg.data);

            // Control response (subscription/unsubscription)
            if ('id' in data) {
              if (data.error) {
                var err = new Error(data.error.msg || 'WebSocket API error');
                err.code = data.error.code;
                if (!resolved) {
                  resolved = true;
                  reject(err);
                } else if (opts.emitStreamErrors) {
                  errorHandler(err);
                }
              } else if (!resolved) {
                resolved = true;
                scheduleRenewal(expirationTime);
                resolve(function (options) {
                  return cleanup(options);
                });
              }
              return;
            }

            // User data event - unwrap if in wrapped format
            var eventData = data;
            if (data.event && _typeof(data.event) === 'object') {
              eventData = data.event;
            }

            // Handle eventStreamTerminated - token expired
            if (eventData.e === 'eventStreamTerminated') {
              cleanup({}, true);
              if (!keepClosed) {
                setTimeout(function () {
                  return _makeStream(true);
                }, 5e3);
              }
              return;
            }
            if (eventData.e && cb) {
              userEventHandler(cb, transform)({
                data: _jsonBigint.default.stringify(eventData)
              });
            }
          };
          w.onerror = function (event) {
            var error = event.error || event.message || new Error('WebSocket error');
            if (opts.emitSocketErrors) {
              errorHandler(typeof error === 'string' ? new Error(error) : error);
            }
          };
          w.onclose = function () {
            if (!keepClosed && resolved) {
              if (renewalTimeout) clearTimeout(renewalTimeout);
              renewalTimeout = null;
              w = null;
              setTimeout(function () {
                return _makeStream(true);
              }, 30e3);
            }
          };
        });
      }).catch(function (err) {
        if (isReconnecting) {
          if (!keepClosed) {
            setTimeout(function () {
              return _makeStream(true);
            }, 30e3);
          }
          if (opts.emitStreamErrors) errorHandler(err);
        } else {
          throw err;
        }
      });
    };
    return _makeStream(false);
  };
};
var user = function user(opts, variator) {
  return function (cb, transform) {
    var _getStreamMethods = getStreamMethods(opts, variator),
      _getStreamMethods2 = _slicedToArray(_getStreamMethods, 3),
      getDataStream = _getStreamMethods2[0],
      keepDataStream = _getStreamMethods2[1],
      closeDataStream = _getStreamMethods2[2];
    var currentListenKey = null;
    var int = null;
    var w = null;
    var keepClosed = false;
    var errorHandler = userErrorHandler(cb, transform);
    var keepAlive = function keepAlive(isReconnecting) {
      if (currentListenKey) {
        keepStreamAlive(keepDataStream, currentListenKey).catch(function (err) {
          closeStream({}, true);
          if (isReconnecting) {
            setTimeout(function () {
              return _makeStream2(true);
            }, 30e3);
          } else {
            _makeStream2(true);
          }
          if (opts.emitStreamErrors) {
            errorHandler(err);
          }
        });
      }
    };
    var closeStream = function closeStream(options, catchErrors) {
      var setKeepClosed = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
      keepClosed = setKeepClosed;
      if (!currentListenKey) {
        return Promise.resolve();
      }
      clearInterval(int);
      var p = closeDataStream({
        listenKey: currentListenKey
      });
      if (catchErrors) {
        p.catch(function (f) {
          return f;
        });
      }
      w.close(1000, 'Close handle was called', _objectSpread({
        keepClosed: true
      }, options));
      currentListenKey = null;
      return p;
    };
    var _makeStream2 = function makeStream(isReconnecting) {
      return !keepClosed && getDataStream().then(function (_ref5) {
        var listenKey = _ref5.listenKey;
        if (keepClosed) {
          return closeDataStream({
            listenKey: listenKey
          }).catch(function (f) {
            return f;
          });
        }
        w = openWebSocket("".concat(variator === 'futures' ? endpoints.futuresPrivate : variator === 'delivery' ? endpoints.delivery : endpoints.base, "/").concat(listenKey));
        w.onmessage = function (msg) {
          return userEventHandler(cb, transform, variator)(msg);
        };
        if (opts.emitSocketOpens) {
          w.onopen = function () {
            return userOpenHandler(cb, transform)();
          };
        }
        if (opts.emitSocketErrors) {
          w.onerror = function (_ref6) {
            var error = _ref6.error;
            return errorHandler(error);
          };
        }
        currentListenKey = listenKey;
        int = setInterval(function () {
          return keepAlive(false);
        }, 50e3);
        keepAlive(true);
        return function (options) {
          return closeStream(options, false, true);
        };
      }).catch(function (err) {
        if (isReconnecting) {
          if (!keepClosed) {
            setTimeout(function () {
              return _makeStream2(true);
            }, 30e3);
          }
          if (opts.emitStreamErrors) {
            errorHandler(err);
          }
        } else {
          throw err;
        }
      });
    };
    return _makeStream2(false);
  };
};
var futuresAllMarkPricesTransform = function futuresAllMarkPricesTransform(m) {
  return m.map(function (x) {
    return {
      eventType: x.e,
      eventTime: x.E,
      symbol: x.s,
      markPrice: x.p,
      indexPrice: x.i,
      settlePrice: x.P,
      fundingRate: x.r,
      nextFundingRate: x.T
    };
  });
};
var _futuresAllMarkPrices = function futuresAllMarkPrices(payload, cb) {
  var transform = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
  var variant = payload.updateSpeed === '1s' ? '!markPrice@arr@1s' : '!markPrice@arr';
  var w = openWebSocket("".concat(endpoints.futuresMarket, "/").concat(variant));
  w.onmessage = function (msg) {
    var arr = _jsonBigint.default.parse(msg.data);
    cb(transform ? futuresAllMarkPricesTransform(arr) : arr);
  };
  return function (options) {
    return w.close(1000, 'Close handle was called', _objectSpread({
      keepClosed: true
    }, options));
  };
};
var _default = exports.default = function _default(opts) {
  if (opts && opts.wsBase) endpoints.base = opts.wsBase;
  if (opts && opts.wsFutures) {
    endpoints.futures = opts.wsFutures;
    endpoints.futuresPublic = opts.wsFutures;
    endpoints.futuresMarket = opts.wsFutures;
    endpoints.futuresPrivate = opts.wsFutures;
  }
  if (opts && opts.wsFuturesPublic) endpoints.futuresPublic = opts.wsFuturesPublic;
  if (opts && opts.wsFuturesMarket) endpoints.futuresMarket = opts.wsFuturesMarket;
  if (opts && opts.wsFuturesPrivate) endpoints.futuresPrivate = opts.wsFuturesPrivate;
  if (opts && opts.wsDelivery) endpoints.delivery = opts.wsDelivery;
  if (opts && opts.proxy) {
    wsOptions.proxy = opts.proxy;
  }
  return {
    depth: depth,
    partialDepth: partialDepth,
    candles: candles,
    trades: trades,
    aggTrades: aggTrades,
    bookTicker: bookTicker,
    ticker: ticker,
    allTickers: allTickers,
    allTickersDeprecated: allTickersDeprecated,
    miniTicker: miniTicker,
    allMiniTickers: allMiniTickers,
    customSubStream: customSubStream,
    user: userWebSocketApi(opts),
    marginUser: marginUserWebSocketApi(opts),
    isolatedMarginUser: function isolatedMarginUser(payload, cb, transform) {
      return marginUserWebSocketApi(opts)(cb, transform, {
        isIsolated: true,
        symbol: payload.symbol,
        validity: payload.validity
      });
    },
    futuresDepth: function futuresDepth(payload, cb, transform) {
      return depth(payload, cb, transform, 'futures');
    },
    deliveryDepth: function deliveryDepth(payload, cb, transform) {
      return depth(payload, cb, transform, 'delivery');
    },
    futuresRpiDepth: futuresRpiDepth,
    futuresPartialDepth: function futuresPartialDepth(payload, cb, transform) {
      return partialDepth(payload, cb, transform, 'futures');
    },
    deliveryPartialDepth: function deliveryPartialDepth(payload, cb, transform) {
      return partialDepth(payload, cb, transform, 'delivery');
    },
    futuresCandles: function futuresCandles(payload, interval, cb, transform) {
      return candles(payload, interval, cb, transform, 'futures');
    },
    deliveryCandles: function deliveryCandles(payload, interval, cb, transform) {
      return candles(payload, interval, cb, transform, 'delivery');
    },
    futuresTicker: function futuresTicker(payload, cb, transform) {
      return ticker(payload, cb, transform, 'futures');
    },
    deliveryTicker: function deliveryTicker(payload, cb, transform) {
      return ticker(payload, cb, transform, 'delivery');
    },
    futuresAllTickers: function futuresAllTickers(cb, transform) {
      return allTickers(cb, transform, 'futures');
    },
    deliveryAllTickers: function deliveryAllTickers(cb, transform) {
      return allTickers(cb, transform, 'delivery');
    },
    futuresAggTrades: function futuresAggTrades(payload, cb, transform) {
      return aggTrades(payload, cb, transform, 'futures');
    },
    deliveryAggTrades: function deliveryAggTrades(payload, cb, transform) {
      return aggTrades(payload, cb, transform, 'delivery');
    },
    futuresLiquidations: futuresLiquidations,
    futuresAllLiquidations: futuresAllLiquidations,
    futuresBookTicker: futuresBookTicker,
    futuresAllBookTickers: futuresAllBookTickers,
    futuresMarkPrice: futuresMarkPrice,
    futuresContinuousCandles: futuresContinuousCandles,
    futuresCompositeIndex: futuresCompositeIndex,
    futuresContractInfo: futuresContractInfo,
    futuresAssetIndex: futuresAssetIndex,
    futuresAllAssetIndex: futuresAllAssetIndex,
    futuresUser: user(opts, 'futures'),
    deliveryUser: user(opts, 'delivery'),
    futuresCustomSubStream: function futuresCustomSubStream(payload, cb) {
      return customSubStream(payload, cb, 'futures');
    },
    deliveryCustomSubStream: function deliveryCustomSubStream(payload, cb) {
      return customSubStream(payload, cb, 'delivery');
    },
    futuresAllMarkPrices: function futuresAllMarkPrices(payload, cb) {
      return _futuresAllMarkPrices(payload, cb);
    }
  };
};
